---
title: "LMR_Report_6.0"
author: "Eugenia Galeota"
date: "13 december 2017"
output: html_document
bibliography: bibliography.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warnings=FALSE)
```


```{r loadinglibs, eval=TRUE, echo=FALSE, message=FALSE, warnings=FALSE}

setwd('/Users/egaleota/Desktop/Methylation/')
library(GenomicRanges)
library(Onassis)
library(data.table)
library(GEOmetadb)
library(compEpiTools)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(org.Hs.eg.db)
library(pheatmap)
library(parallel)
library(magrittr)
```

### Introduction

Methylation of cytosines in metazoan genomes adds epigenetic information on DNA without changing the genetic information. Mehtylation is thought to locally influence the regulatory potential of DNA, leading to changes in gene expression during development and disease. High througput sequencing enables the quantification of DNA methylation genome-wide at base-pair and single-molecule resolution in unprecedented breadth and detail. Using whole-genome bisulphite sequencing (WGBS) it is also possible to identify a novel epigenome feature defined by localized reduced levels of methylation (LMRs). 
Different studies contrasting LMRs with maps of histone modifications, DNase I hypersensitivity and several DNA-binding factors revealed that they are distal regulatory regions. 
LMRs form dynamically during differentiation driven by cell-type-specific factors, revealing a continuous crosstalk between DNA-binding factors and local DNA methylation.

### Loading identified Low Methylated regions and the metadata of Marmalaid database

```{r loadingdata, eval=TRUE, echo=TRUE}
load("./data/Infi_LMRs.Rdata")
marmalaid_annotations <- readRDS("./data/marmalaid_annotations.rds")
lengths <- lapply(Infi_LMRs, length)
Infi_LMRs <- Infi_LMRs[!names(Infi_LMRs) %in% names(which(lengths==0))]
```

As starting data we have Marmal-aid annotations and Infi_LMRs identified by a procedure described in Onassis paper. In the following plot we show the distribution of number of LMRs int the `r length(Infi_LMRs)` total GEO and TCGA samples available in Marmal-aid. 

```{r plot_number_of_lmrs_for_each_sample, eval=TRUE, echo=FALSE}
plot(density(unlist(lapply(Infi_LMRs, length))), main='Number of LMRs for each sample', xlab='LMRs', col='red', lwd=3)
summary(unlist(lapply(Infi_LMRs, length)))
```

The number of LMRs ranges from a few thousands to many thousands for a given sample. 

### Semantic annotation of Marmal-aid samples' metadata with concepts from biomedical ontologies using Onassis

We decided to annotate Marmal-aid metadata in order to map the annotated terms to Disease ontology and Cell Line ontology concepts. For this purpose we have used Onassis. 
The ontologies that we want to use for the annotation can be retrieved from the obolibrary website. Onassis automatically creates the dictionary instances for CL (Cell Line Ontology) and DOID (Human disease Ontology). Since DO does not contain a concept to identify 'Healthy' individuals we also added some code to find 'HEALTHY' entities.


```{r tissuediseaseannot, echo=TRUE, eval=FALSE}
cell_obo <- 'http://purl.obolibrary.org/obo/cl.obo'
#Mapping Marmalaid annotations to CL cell lines and DO diseases
marmalaid_annotations$Id <- gsub("-", "_", marmalaid_annotations$Id)

# 'secretion', 'cell and encapsulating structures', 'Homo sapiens'
#Definition of terms to exclude
terms_to_exclude <- c('tissue' , 'mesoderm', 'ectoderm', 'endoderm', 'cell', )

#Annotation of the field LINEAGE with CL
lineages <- unique(marmalaid_annotations$LINEAGE)[which(!is.na(unique(marmalaid_annotations$LINEAGE)))]
lineages <- data.frame(cbind(paste0('lineage_', seq(1,length(lineages), 1)), lineages))
system.time(marmalaid_lineage_entities_cell <- annot(lineages, 'OBO', cell_obo))

marmalaid_lineage_entities_cell <- filterconcepts(marmalaid_lineage_entities_cell, terms_to_exclude)
cell_obo <- file.path(getwd(), 'cl.obo')

#Annotation of the field TISSUE with CL
marmalaid_annotations$TISSUE <- gsub("^\\s+|\\s+$", "", marmalaid_annotations$TISSUE)
tissues <- unique(marmalaid_annotations$TISSUE)[which(!is.na(unique(marmalaid_annotations$TISSUE)))]
tissues <- data.frame(cbind(paste0('tissue_', seq(1,length(tissues), 1)), tissues))
system.time(marmalaid_tissue_entities_cell <- annot(tissues, 'OBO', cell_obo))
marmalaid_tissue_entities_cell <- filterconcepts(marmalaid_tissue_entities_cell, terms_to_exclude)

#Annotation of the field TISSUE_SUBTYPE with CL
marmalaid_annotations$TISSUE_SUBTYPE <- gsub("^\\s+|\\s+$", "", marmalaid_annotations$TISSUE_SUBTYPE)
tissue_subtypes <- unique(marmalaid_annotations$TISSUE_SUBTYPE)[which(!is.na(unique(marmalaid_annotations$TISSUE_SUBTYPE)))]
tissue_subtypes <- data.frame(cbind(paste0('tissuesub_', seq(1,length(tissue_subtypes), 1)), tissue_subtypes))
system.time(marmalaid_tissue_sub_entities_cell <- annot(tissue_subtypes, 'OBO', cell_obo))
marmalaid_tissue_sub_entities_cell <- filterconcepts(marmalaid_tissue_sub_entities_cell, terms_to_exclude)



#Creation of tables to evaluate the performaces of Onassis
lineage_eval <- merge(lineages, unique(entities(marmalaid_lineage_entities_cell)[,c('sample_id', 'term_name', 'term_url')]), by.x='V1', by.y='sample_id', all.x=TRUE)
colnames(lineage_eval) <- c('Term_ID', 'Marmalaid_Annotation', 'Onassis', 'term_url')

tissue_eval <- merge(tissues, unique(entities(marmalaid_tissue_entities_cell)[,c('sample_id', 'term_name', 'term_url')]), by.x='V1', by.y='sample_id', all.x=TRUE)
colnames(tissue_eval) <- c('Term_ID', 'Marmalaid_Annotation', 'Onassis', 'term_url') 

tissue_sub_eval <- merge(tissue_subtypes, unique(entities(marmalaid_tissue_sub_entities_cell)[,c('sample_id', 'term_name', 'term_url')]), by.x='V1', by.y='sample_id', all.x=TRUE)
colnames(tissue_sub_eval) <- c('Term_ID', 'Marmalaid_Annotation', 'Onassis', 'term_url')

#Merging  lineage cell entitites 
marmalaid_annotations <- merge(marmalaid_annotations, unique(lineage_eval[,c('Marmalaid_Annotation', 'term_url', 'Onassis')]), by.x='LINEAGE', by.y='Marmalaid_Annotation', all.x=TRUE)
colnames(marmalaid_annotations)[(ncol(marmalaid_annotations)-1):ncol(marmalaid_annotations)] <- c('lineage_cell_id', 'lineage_cell_name')

#Merging tissue cell entities
marmalaid_annotations <- merge(marmalaid_annotations, unique(tissue_eval[,c('Marmalaid_Annotation', 'term_url', 'Onassis')]), by.x='TISSUE', by.y='Marmalaid_Annotation', all.x=TRUE)
colnames(marmalaid_annotations)[(ncol(marmalaid_annotations)-1):ncol(marmalaid_annotations)] <- c('tissue_cell_id', 'tissue_cell_name')

#Merging tissue_subtype cell entities
marmalaid_annotations <- merge(marmalaid_annotations, unique(tissue_sub_eval[,c('Marmalaid_Annotation', 'term_url', 'Onassis')]), by.x='TISSUE_SUBTYPE', by.y='Marmalaid_Annotation', all.x=TRUE)
colnames(marmalaid_annotations)[(ncol(marmalaid_annotations)-1):ncol(marmalaid_annotations)] <- c('tissue_sub_cell_id', 'tissue_sub_cell_name')


saveRDS(marmalaid_annotations, file='./data/onassisresults/marmalaid_annotations_uncollapsed_tissue.rds')

#Creating a data table with marmalaid_annotations
setDT(marmalaid_annotations)

#Collapsing the annotations in single annotation lines by sample Id
marmalaid_annotations_collapsed <- marmalaid_annotations[, lapply(.SD, function(x) toString(unique(x[order(x)]))), by=Id]
marmalaid_annotations_collapsed$lineage_cell_id <- gsub("^\\,\\s+", "", marmalaid_annotations_collapsed$lineage_cell_id)
marmalaid_annotations_collapsed$lineage_cell_name <- gsub("^\\,\\s+", "", marmalaid_annotations_collapsed$lineage_cell_name)

saveRDS(marmalaid_annotations_collapsed, file='./data/onassisresults/marmalaid_annotations_collapsed_tissue.rds')
```

```{r loading_tissues, echo=FALSE, eval=TRUE}
marmalaid_annotations_collapsed <- readRDS('./data/onassisresults/marmalaid_annotations_collapsed_tissue.rds')
```

With the following code we computed the number of GEO samples for each semantic group.

```{r NumberOfSamplesForEachAnnotation, echo=TRUE, eval=TRUE}
#Counting the number of samples associated to each semantic annotation class
unique_marmalaid_annotations_names <- marmalaid_annotations_collapsed[, .N, by=list(lineage_cell_name, tissue_cell_name, tissue_sub_cell_name)]
```

```{r NumberOfSamplesForEachAnnotationPlot, echo=FALSE, eval=TRUE}
#Counting the number of samples associated to each semantic annotation class
boxplot(unique_marmalaid_annotations_names$N, main='Number of samples for each combination of annotations', col='violet')
summary(unique_marmalaid_annotations_names$N)
```

#Creating the table of annotated LMRs

We summarized the annotations of a given GEO sample by concatenating the unique annotation ids and annotation names. In this way, each sample has a unique annotation, defined as "semantic class" or "semantic set". Each semantic set can include multiple concepts from Cell Line Ontology.  

```{r concatenation_of_annotations, echo=TRUE, eval=FALSE}

marmalaid_annotations_collapsed2 <- as.data.frame.matrix(marmalaid_annotations_collapsed)

#Creation of a data frame where each row contains the annotation ids and annotation names pasted for the first element in the row, that is a geo sample
trim <- function (x) gsub("^\\s+|\\s+$", "", x)
m_a_c <- t(apply(marmalaid_annotations_collapsed2, 1, function(annotation) {
  id <- annotation[1]
  annot_ids <- unique(annotation[c(14, 16, 18)])
  annot_names <- unique(annotation[c(15, 17, 19)])
  annot_ids <- paste(unique(annot_ids[which(!annot_ids=='NA' & !annot_ids==""  & !is.na(annot_ids))]), collapse=',')
  annot_names <- paste(unique(annot_names[which(annot_names!='NA' & annot_names!="" & !is.na(annot_names))]), collapse=',')
  annot_ids <- strsplit(annot_ids, ',')
  annot_names <- strsplit(annot_names, ',')
  annot_ids <- sapply(annot_ids, trim)
  annot_names <- sapply(annot_names, trim)
  annot_ids <- unique(annot_ids)
  annot_names <- unique(annot_names)
  annot_ids <- paste(annot_ids, collapse=',')
  annot_names <- paste(annot_names, collapse=',')
  c(id, annot_ids, annot_names)
}))

m_a_c[m_a_c[,3]=='character(0)', 2] <- ""
m_a_c[m_a_c[,3]=='character(0)', 3] <- ""

m_a_c[,1] <- gsub('_', '-', m_a_c[,1])
rownames(m_a_c) <- m_a_c[,1]
m_a_c <- as.data.frame(m_a_c)
colnames(m_a_c) <- c('Sample_ID', 'Annot_IDs', 'Annot_Names' )
saveRDS(m_a_c, './data/m_a_c_tissues.rds')
```

```{r sample_to_annotations_table_loading, echo=FALSE, eval=TRUE}
m_a_c_tissue <- readRDS('./data/m_a_c_tissues.rds')
```

The following table shows a sample of the summarized annotations. 

```{r headannotations, echo=FALSE, eval=TRUE}
head(m_a_c_tissue[sample(nrow(m_a_c_tissue), 20), ])
m_a_c <- m_a_c_tissue
```

We obtained `r length(unique(m_a_c$Annot_IDs[which(!m_a_c$Annot_IDs=="")]))` unique annotation sets for `r length(unique(m_a_c$Sample_ID))` samples.

###Building the semantic similarity network between semantic annotation sets

Semantic similarity between  eache couple of semantic sets has been obtained using the similarity functions provided by Onassis. For each couple of semantic sets, the semantic similarity can be computed as the groupwise semantic similarity between cell line concepts of the CL ontology of each of the two sets. To compute the groupwise semantic similarity we set Onassis to use the "Best Match Average strategy" to group LIN pairwise semantic similarities [@Galeota03052016] 

```{r semanticsimilaritymatrix, echo=TRUE, eval=FALSE}

#Firstly we create a new Onassis object to contain our merged annotations
onassis <- new('Onassis')
dict(onassis) <- dict(marmalaid_lineage_entities_cell)
colnames(m_a_c_tissue) <- c('sample_id', 'term_url', 'term_name')

m_a_c_tissue$term_id <- gsub('http://purl.obolibrary.org/obo/','',m_a_c_tissue$term_url )
entities(onassis) <- m_a_c_tissue[which(m_a_c_tissue$term_url!=''),]

onassis <- sim(onassis)
semantic_net_matrix <- simil(onassis)
saveRDS(semantic_net_matrix, file='./data/semantic_net_matrix_tissues.rds')
```

```{r semanticsimilaritymatrix_load, echo=FALSE, eval=TRUE}
semantic_net_matrix <- readRDS('./data/semantic_net_matrix_tissues.rds')
```

Pairwise semantic similarities between semantic classes were used to create a semantic similarity matrix, that in turn has been used to create a complete weighted network, where nodes are the semantic sets and as edges are their semantic similarity values. 
Semantic sets can be clustered based on a distance calculated as "1 - groupwise semantic similarities" between each couple of semantic sets.

<out.extra='angle=90', cache=TRUE, echo=FALSE, message=FALSE, warning=FALSE, comment=NA, eval=TRUE, results='asis'>

```{r semantic_net_graph, echo=TRUE, eval=TRUE, fig.height=40, fig.width=10}

#The following lines of codes are used to plot the hierarchical clustering of annotations based on the distance obained by 1 - lin semantic similarity
#png('./images/Semantic_dendrogram.png' , width=5000, height=7000, res=350)
op <- par(mar = c(2,2,2,30) + 0.1)

semantic_distance <- 1 - as.matrix(semantic_net_matrix)
plot(as.dendrogram(hclust(d=as.dist(semantic_distance))), main='Hierarchical clustering of CL (Cell Ontology) semantic sets', col.axis='#F38630', col.lab="#7c8071", lwd=3, sub='', las=1, horiz=TRUE)
#dev.off()
```

In the following chart we show the distribution of weights in the semantic network.

```{r semantic_net_graph_stats, echo=TRUE, eval=TRUE}
hist(semantic_net_matrix[upper.tri(semantic_net_matrix)], main='Weights frequency distribution of the semantic network', xlab='Semantic similarity', col='red')
```

###Simplifying the semantic net matrix###

We decided to collapse semantically close annotation sets by cutting the hierarchical clustering tree and merging similar sets. In more detail, the semantic sets having semantic similarity higher than 0.8 were merged into a single semantic set, named with the three most frequent single concepts of the list of unique concepts obtained merging the semantic sets. 

```{r collapsing_semantic_states, echo=TRUE, eval=FALSE}

collapsed_onassis <- collapse(onassis, 0.8)
saveRDS(collapsed_onassis, file='./data/collapsed_onassis.rds')

```

```{r loading_m_a_c_collapsed, echo=FALSE, eval=TRUE}
collapsed_onassis <- readRDS('./data/collapsed_onassis.rds')
m_a_c_tissue_collapsed <- entities(collapsed_onassis)
semantic_net_matrix <- simil(collapsed_onassis)
```

We finally obtained `r length(unique(m_a_c_tissue_collapsed$term_id))` semantic annotation sets, from which we built again the semantic similarity matrix.


```{r semantic_net_graph_collapsed, echo=TRUE, eval=FALSE, fig.height=40, fig.width=10}

semantic_net_matrix <- simil(collapsed_onassis)
semantic_distance <- as.dist(1- semantic_net_matrix)

#png('./images/Semantic_dendrogram_collapsed.png' , width=2700, height=5000, res=350)
op <- par(mar = c(2,2,2,30) + 0.1)

#Getting the labels 
dendr <- as.dendrogram(hclust(d=semantic_distance))
lab <- labels(dendr)
occurrences <- sapply(strsplit(lab, "\\((?=[^\\(]+$)", perl=TRUE), '[[', 2)
occurrences <- as.numeric(gsub(')', '', occurrences))
label_sizes <- cut(occurrences, quantile(occurrences), include.lowest = TRUE, labels=c(0.7, 0.8, 0.9, 1))
dendr2 <- dendr %>% set("labels", lab) %>% set("labels_cex", as.numeric(as.vector(label_sizes)))
dendr2 %>% plot(col.axis='#F38630', col.lab="#7c8071", horiz=TRUE) #lwd=3, sub='', las=1, 
#dev.off()
```


### Building the bipartite network of LMRs and Semantic Annotation Sets

LMRs in distal regions, e.g. the regions that are neither in gene bodies nor in promoters are usually associated with enhancers that drive the expression of cell identity genes. We decided to focus on those elements, to find LMRs that are tissue/cell line specific. We filtered out the LMRs in gene bodies and promoter regions from our set of LMRs. 

```{r listofDistalLMRs, echo=TRUE, eval=FALSE}
# Focusing on distal elements (This piece of code uses parallel programming)
#library(parallel)
#library(compEpiTools)
#library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
#load('./data/Infi_LMRs.Rdata')
#library(org.Hs.eg.db)
EG2GS <- org.Hs.eg.db

downstream = 1000
upstream = 2000
gb <- transcripts(txdb, columns=c('tx_id','tx_name','gene_id'))
suppressWarnings(prom <- promoters(txdb, upstream=upstream, downstream=downstream, 
                                   columns=c('tx_id','tx_name','gene_id')))
names(prom) <- prom$gene_id
names(gb) <- gb$gene_id
inds <- which(start(prom)<0)
if(length(inds)>0) start(prom)[inds]=1
ind <- which(duplicated(gb$tx_name)==TRUE)
if(length(ind) > 0) {
  gb <- gb[-c(ind)]
  prom <- prom[-c(ind)]
}

# cutting promoter regions downstream the TSS from genebody regions
shortinds <- which(width(gb) <= downstream)
gbNotProm <- gb[-shortinds]
plusStr <- which(as.logical(strand(gbNotProm) == '+'))
minusStr <- which(as.logical(strand(gbNotProm) == '-'))
start(gbNotProm[plusStr]) <- start(gbNotProm[plusStr]) + downstream
end(gbNotProm[minusStr]) <- end(gbNotProm[minusStr]) - downstream


system.time(Infi_LMRs_intergenic <- mclapply(Infi_LMRs, function(sample_granges){
  gbInGR <- findOverlaps(query=gbNotProm, subject=GRmidpoint(sample_granges),
                         type='any', select='all')
  promInGR <- findOverlaps(query=prom, subject=GRmidpoint(sample_granges),
                           type='any', select='all')
  if(length(subjectHits(gbInGR))>0)
    sample_granges <- sample_granges[-subjectHits(gbInGR)]
  if(length(subjectHits(promInGR))>0)
    sample_granges <- sample_granges[-subjectHits(promInGR)]
  sample_granges
}, mc.cores=12))

saveRDS(Infi_LMRs_intergenic, file='./data/Infi_LMRs_intergenic.rds')
```

```{r loadingDISTALS, echo=FALSE, eval=TRUE}
Infi_LMRs_intergenic <- readRDS('./data/Infi_LMRs_intergenic.rds')
```

From the table of Onassis tissue annotations, we created a new table, by replacing the GEO identifiers with the list of distal LMRs found in that sample. For each couple LMR - annotation set we also stored the number of samples supporting that evidence. Then for each semantic set having at least two samples we considered only the LMRs that were in at least the 70% of the samples in the semantic set. 


```{r filtering_the_tissue_with_70_percent, echo=TRUE, eval=FALSE}

unique_tissues <- unique(m_a_c_tissue_collapsed$term_name)

#Finding the list of LMRs associated to each GSM for each disease class in each tissue 
tissue_2_lmrs <- lapply(unique_tissues, function(tissue_name){
  gsms <- unique(m_a_c_tissue_collapsed$sample_id[which(m_a_c_tissue_collapsed$term_name==tissue_name)])
  lmrs <- unlist(lapply(Infi_LMRs_intergenic[gsms], as.character))
})
names(tissue_2_lmrs) <- unique_tissues

#Filtering the LMRs in the semantic sets, by keeping only those available in at least the 70% of the samples 
f_tissue_2_lmrs <- lapply(names(tissue_2_lmrs), function(tissue_name){
  num_sam <- length(unique(m_a_c_tissue_collapsed$sample_id[which(m_a_c$term_name==tissue_name)]))
  percentage <- round(num_sam * 70 / 100)
  tabella <- table(tissue_2_lmrs[[tissue_name]])
  lmrs <- names(which(tabella >= percentage))
})

names(f_tissue_2_lmrs) <- names(tissue_2_lmrs)
saveRDS(f_tissue_2_lmrs, file='./data/tissue_2_lmrs_70_percent.rds')
```


```{r computingEdgeList_collapsed_tissue, echo=TRUE, eval=FALSE}
# This code returns an edge list with LMRs and semantic sets
# Each edge is weighted by the number of samples annotated with that semantic set and 
# containing that LMR

Infi_LMRs_intergenic <- Infi_LMRs_intergenic[which(names(Infi_LMRs_intergenic) %in% m_a_c_tissue_collapsed$sample_id)]
Infi_LMRS_string  <- lapply(Infi_LMRs_intergenic, function(granges) as.character(granges, ignore.strand=FALSE)) 
names(Infi_LMRS_string) <- names(Infi_LMRs_intergenic)

rownames(m_a_c_tissue_collapsed) <- m_a_c_tissue_collapsed$sample_id

#Creating a list of annotations for all the lmrs in all the samples
system.time(edges_lists <- sapply(names(Infi_LMRS_string), function(sample){
  lmrs_in_sample <- Infi_LMRS_string[[sample]]
  cbind(lmrs_in_sample, rep(as.character(m_a_c_tissue_collapsed[sample, c('term_name')]), length(lmrs_in_sample)) )
}))

#Creating a unique list of edges for a bipartite network
edge_list <- do.call("rbind", edges_lists)
edge_list <- setDT(as.data.frame(edge_list))
colnames(edge_list)[2] <- 'annotation'
#saveRDS(edge_list, file='./data/edge_list_tissue_collapsed.rds')
system.time(unique_edge_list <- edge_list[, .N, by = .(lmrs_in_sample, annotation)])
#saveRDS(unique_edge_list, file='./data/unique_edge_list_tissue_collapsed.rds')

#Filtering the LMRs that that are not supported by 70% of the samples
unique_edge_list_collapsed_filtered_70 <- unique_edge_list[which(unique_edge_list$lmrs_in_sample %in% unique(unlist(f_tissue_2_lmrs))),]
saveRDS(unique_edge_list_collapsed_filtered_70, './data/unique_edge_list_tissue_collapsed_70_percent.rds')
```

```{r loading_edge_list, echo=FALSE, eval=TRUE}
unique_edge_list_collapsed <- readRDS('./data/unique_edge_list_tissue_collapsed_70_percent.rds')
```

The list of unique LMRs - semantic set annotations has a total of `r nrow(unique_edge_list_collapsed)` associations for `r length(unique(unlist(unique_edge_list_collapsed[,1])))` LMRs and `r length(unique(unlist(unique_edge_list_collapsed[,2])))` tissue semantic sets.

```{r ENTIRE_BIPARTITE_MATRIX_collaps_70_percent, echo=TRUE, eval=FALSE}

unique_edge_list <- as.data.frame.matrix(unique_edge_list_collapsed)
ptm <- proc.time()
bipartite_matrix <-  matrix(0, length(unique(unique_edge_list$annotation)), length(unique(unique_edge_list$lmrs_in_sample)))
rownames(bipartite_matrix) <- unique(as.character(as.vector(unique_edge_list$annotation)))
colnames(bipartite_matrix) <- unique(as.character(as.vector(unique_edge_list$lmrs_in_sample)))

system.time(apply(unique_edge_list, 1, function(edge){
  annot <- as.character(as.vector(edge[2]))
  lmr <- as.character(as.vector(edge[1]))
  bipartite_matrix[annot, lmr] <<- as.numeric(edge[3])
}))

rownames(bipartite_matrix) <- m_a_c_tissue_collapsed$short_label[match(rownames(bipartite_matrix), m_a_c_tissue_collapsed$term_name)]
saveRDS(bipartite_matrix, file='./data/bipartite_matrix_tissue_collapsed_70_percent.rds')
print(proc.time() - ptm)
```


```{r loading_bipartite_matrix, echo=FALSE, eval=TRUE}
bipartite_matrix_tissue <- readRDS('./data/bipartite_matrix_tissue_collapsed_70_percent.rds')
```


```{r heatmap_of_bipartite_matrices_2, echo=FALSE, eval=TRUE}

bipartite_matrix_corrected <- bipartite_matrix_tissue
bipartite_matrix_corrected <- bipartite_matrix_corrected / rowSums(bipartite_matrix_corrected) 
rownames(bipartite_matrix_corrected) <- gsub('\\]\\]', '\\]', rownames(bipartite_matrix_corrected))
rownames(bipartite_matrix_corrected) <- gsub('\\[\\[', '\\[', rownames(bipartite_matrix_corrected))
colnames(semantic_net_matrix) <- rownames(semantic_net_matrix) <- gsub('\\[\\[', '\\[', rownames(bipartite_matrix_corrected))
colnames(semantic_net_matrix) <- rownames(semantic_net_matrix) <- gsub('\\]\\]', '\\]', rownames(bipartite_matrix_corrected))

lmrs_distances <- dist(t(bipartite_matrix_corrected))
hclus_lmrs_distances <- hclust(lmrs_distances)

semantic_distance <- 1 - as.matrix(semantic_net_matrix)
hclus_semantic_sets_distances <- hclust(d=as.dist(semantic_distance))



#bipartite_matrix_corrected[which(bipartite_matrix_corrected >100)] <- 100
bipartite_matrix_corrected_ordered <- bipartite_matrix_corrected[hclus_semantic_sets_distances$labels,]

pheatmap(bipartite_matrix_corrected_ordered[,hclus_lmrs_distances$order], cluster_rows=hclus_semantic_sets_distances, cluster_cols=F, #filename='./images/bipartite_matrix.png',
         show_colnames=FALSE, show_rownames = TRUE, width = 20, height=20, fontsize_row = 15)
```


In the following table we show a sample of the associations lmrs-annotations with the number of samples. 

```{r headunique, echo=FALSE, eval=TRUE}
head(unique_edge_list_collapsed)
```

The previous heatmap shows that most edges are supported by few evidences, while there are some LMRs and their annotations appearing in a very large number of samples. In the following plot we show the distribution of the number of samples in the semantic classes

```{r densityPlotOfLMRs, echo=FALSE, eval=TRUE}
summary(unique_edge_list_collapsed$N)
boxplot(as.numeric(as.vector(unique_edge_list_collapsed$N)), outline=FALSE, main='Distribution of the num of samples in LMRs-Semantic sets', xlab='Samples', col='green')
```

In the following list we show the most common annotations for the top rich 5% couples of annotations and lmrs.

```{r listOfTopAnnotations, echo=FALSE, eval=FALSE}
m_a_c <- readRDS('./data/m_a_c_collapsed_semantic_sets.rds')
unique(m_a_c[which(m_a_c$nuovi_ids %in% unique(unique_edge_list[N > quantile(unique_edge_list$N, probs=c(0.95))]$annotation)), c('Term_Name')])
```

By looking at the distribution of the number of samples in which each LMR appears, without considering the semantic annotation of the samples, we have an idea of the specificity/generality of LMRs within the samples. 

```{r Nnumber_of_lmr_for_samples, echo=TRUE, eval=FALSE}

number_of_samples_for_each_lmr <- edge_list[, .N, by=.(lmrs_in_sample)]
#boxplot(number_of_samples_for_each_lmr$N, main='Number of samples for LMRs', xlab='Samples', col='orange')
plot(density(number_of_samples_for_each_lmr$N), main='Number of samples for each LMR', xlab='Samples', col='orange')
```

The bimodal distribution suggests the presence of LMRs that appear in almost all samples, while most of the LMRs are located in a small percentage of the samples, suggesting, as supported also by the literature, that distal LMRs could be tissue/cell line specific. 
To have insights in the tissue specificity we built a network between LMRs and  Semantic Annotation Sets. Our network is a two-mode network, where each edge connects a given LMR i with a semantic annotation set j weighted by the number of samples containing the LMR i and annotated with the semantic annotation set j.The network has been built with igraph.

```{r bipartite_network_construction, echo=TRUE, eval=FALSE}

setDT(unique_edge_list)
#unique_edge_list_filtered <- unique_edge_list[N > 100 & N < 800]
#Lets' focus only on LMRs in the chromosome one

unique_edge_list_filtered <- unique_edge_list
#unique_edge_list_filtered <- unique_edge_list[grep('chr1:', lmrs_in_sample)]
setDT(unique_edge_list_filtered)

#recomputing the number of samples for each LMR
number_of_samples_for_each_lmr <- unique_edge_list_filtered[, .N, by=.(lmrs_in_sample)]

unique_edge_list_filtered <- as.data.frame.matrix(unique_edge_list_filtered)

g <- graph.empty()

#Adding LMRs as nodes of the first type for the bipartite network
g <- add.vertices(g,nv=length(unique(unique_edge_list_filtered$lmrs_in_sample)),attr=list(name=unique(unique_edge_list_filtered$lmrs_in_sample)),       type=rep(TRUE,length(unique(unique_edge_list_filtered$lmrs_in_sample))))

#Adding semantic annotations as nodes of the second type in the bipartite network
g <- add.vertices(g,nv=length(unique(unique_edge_list_filtered$annotation)),attr=list(name=unique(unique_edge_list_filtered$annotation)),                          type=rep(FALSE,length(unique(unique_edge_list_filtered$annotation))))

#Adding the edges to the network, using as weight the number of samples 
edgeListVec <- as.vector(t(as.matrix(unique_edge_list_filtered[,c(1:2)])))
g <- add.edges(g, edgeListVec,  attr=list(samples=unique_edge_list_filtered[,3]))
E(g)$weight = unique_edge_list_filtered[,3]

#Just a check to verify that g is bipartite
#is.bipartite(g)
```


```{r bipartite_degree_distribution, echo=TRUE, eval=FALSE}
png('./images/figure_4.png', height=1000, width=2000, res=300)
deg<-igraph::degree(g)[V(g)$type==TRUE]
summary(deg)

plot(density(deg), xlab='LMRs degree', main='LMRs degree (number of semantic sets) distribution', col='red', lwd=3)
dev.off()
```

The out-degree distribution of the network shows a bimodal behaviour, reflecting the distribution of the LMRs in samples, with most of the LMRs appearing in a lower number of semantic annotation sets and few LMRs appearing in almost all semantic annotation sets.   

```{r annotations_overlap, echo=TRUE, eval=FALSE}
library(foreach)
library(doParallel)
library('Matrix')
cores=12
setwd('/data/BA/egaleota/methylation')
cl <- makeCluster(cores) #not to overload your computer
registerDoParallel(cl)
unique_edge_list <- readRDS('./datasets/unique_edge_list_tissue_collapsed.rds')
lmrs <- unique(as.character(as.vector(unique_edge_list$lmrs_in_sample)))
annotations <- unique(unique_edge_list$annotation)
number_of_annotations <- length(annotations) - 1
chis_m <- matrix(NA, nrow=length(annotations), ncol=length(annotations))
pval_m <- matrix(NA, nrow=length(annotations), ncol=length(annotations))
overlaps_m <- matrix(NA, nrow=length(annotations), ncol=length(annotations))
rownames(chis_m) <- colnames(chis_m) <- annotations
rownames(pval_m) <- colnames(pval_m) <- annotations
rownames(overlaps_m) <- colnames(overlaps_m) <- annotations

for(i in 1:number_of_annotations) {
  sclass1 <- as.character(as.vector(annotations[i]))
  lmrs1 <- unique(unique_edge_list$lmrs_in_sample[which(unique_edge_list$annotation==sclass1)])
  j = i + 1
  for(k in j:length(annotations)){
    print(paste0(i, '---', k))
    sclass2 <- annotations[k]   
    lmrs2 <- unique(unique_edge_list$lmrs_in_sample[which(unique_edge_list$annotation==sclass2)])
    lmrs1_lmrs2 <- length(intersect(lmrs1, lmrs2)) 
    lmrs1_not_lmrs2 <- length(lmrs1[which(!lmrs1 %in% lmrs2)])
    lmrs2_not_lmrs1 <- length(lmrs2[which(!lmrs2 %in% lmrs1)]) 
    lmrs_universe <- length(lmrs[which(!lmrs %in% lmrs1 & !lmrs %in% lmrs2)]) 
    lmrs_contingency <- matrix(c(lmrs1_lmrs2, lmrs1_not_lmrs2, lmrs2_not_lmrs1, lmrs_universe), ncol=2, byrow=TRUE)
    chis <- fisher.test(lmrs_contingency)
    odds <- chis$estimate
    pvalues <- chis$p.value
    overlaps <- lmrs1_lmrs2 / length(union(lmrs1, lmrs2))
    chis_m[i, k] <- chis_m[k, i] <- odds
    pval_m[i, k] <- pval_m[k, i] <- pvalues
    overlaps_m[i, k] <- overlaps_m[k, i] <- overlaps
  }
}

matrix_list <- list(chis_m, pval_m, overlaps_m)
saveRDS(matrix_list, file='./datasets/chisquares_tissue.rds')
```

```{r annotations_overlap_70_percent, echo=TRUE, eval=FALSE}
library(foreach)
library(doParallel)
library('Matrix')
cores=12
setwd('/data/BA/egaleota/methylation')
cl <- makeCluster(cores) #not to overload your computer
registerDoParallel(cl)
unique_edge_list <- readRDS('./datasets/unique_edge_list_tissue_collapsed_70_percent.rds')
shannon_specific <- readRDS('./datasets/most_specific_lmrs_shannon_70_percent.rds')
unique_edge_list <- unique_edge_list[which(unique_edge_list$lmrs_in_sample %in% shannon_specific),]
lmrs <- unique(as.character(as.vector(unique_edge_list$lmrs_in_sample)))
annotations <- unique(unique_edge_list$annotation)
number_of_annotations <- length(annotations) - 1
chis_m <- matrix(NA, nrow=length(annotations), ncol=length(annotations))
pval_m <- matrix(NA, nrow=length(annotations), ncol=length(annotations))
overlaps_m <- matrix(NA, nrow=length(annotations), ncol=length(annotations))
rownames(chis_m) <- colnames(chis_m) <- annotations
rownames(pval_m) <- colnames(pval_m) <- annotations
rownames(overlaps_m) <- colnames(overlaps_m) <- annotations

for(i in 1:number_of_annotations) {
  sclass1 <- as.character(as.vector(annotations[i]))
  lmrs1 <- unique(unique_edge_list$lmrs_in_sample[which(unique_edge_list$annotation==sclass1)])
  j = i + 1
  for(k in j:length(annotations)){
    print(paste0(i, '---', k))
    sclass2 <- annotations[k]   
    lmrs2 <- unique(unique_edge_list$lmrs_in_sample[which(unique_edge_list$annotation==sclass2)])
    lmrs1_lmrs2 <- length(intersect(lmrs1, lmrs2)) 
    lmrs1_not_lmrs2 <- length(lmrs1[which(!lmrs1 %in% lmrs2)])
    lmrs2_not_lmrs1 <- length(lmrs2[which(!lmrs2 %in% lmrs1)]) 
    lmrs_universe <- length(lmrs[which(!lmrs %in% lmrs1 & !lmrs %in% lmrs2)]) 
    lmrs_contingency <- matrix(c(lmrs1_lmrs2, lmrs1_not_lmrs2, lmrs2_not_lmrs1, lmrs_universe), ncol=2, byrow=TRUE)
    chis <- fisher.test(lmrs_contingency)
    odds <- chis$estimate
    pvalues <- chis$p.value
    overlaps <- lmrs1_lmrs2 / length(union(lmrs1, lmrs2))
    chis_m[i, k] <- chis_m[k, i] <- odds
    pval_m[i, k] <- pval_m[k, i] <- pvalues
    overlaps_m[i, k] <- overlaps_m[k, i] <- overlaps
  }
}

matrix_list <- list(chis_m, pval_m, overlaps_m)
saveRDS(matrix_list, file='./datasets/chisquares_tissue_70_percent_specific.rds')
```




The odds ratio Fisher's statistics measures the significance of overlap of LMRs within two Semantic Annotations sets. For each couple of semantic annotation sets we computed the number of LMRs in the intersection of the two sets, the number of LMRs present only in the first semantic annotation set, the number of LMRs present only in the second set and the number of LMRs that are neither in the first, nor in the second annotation set. 
In the following plots we show the distribution of the odds ratio values in increasing intervals of semantic similarity between semantic sets. Odds ratios include 0 values either for two semantic sets sharing 0 LMRs and for two semantic sets containing the universe of LMRS. Odds ratios having Inf values are reported instead for annotation sets where one of the two is a subset of the other. 


```{r overlap, echo=FALSE, eval=FALSE, fig.height=22}
png('./images/figure_5.png', height=4000, width=4000, res=300)

overlap_matrices <- readRDS('./data/ieo_cluster/chisquares_tissue.rds')
semantic_net_matrix <- readRDS('./data/semantic_net_matrix_tissues_collapsed.rds')

overlap_matrices[[1]] <- overlap_matrices[[1]][rownames(semantic_net_matrix), colnames(semantic_net_matrix)]
overlap_matrices[[2]] <- overlap_matrices[[2]][rownames(semantic_net_matrix), colnames(semantic_net_matrix)]
overlap_matrices[[3]] <- overlap_matrices[[3]][rownames(semantic_net_matrix), colnames(semantic_net_matrix)]
overlap_matrices[[1]] <- as.numeric(as.vector(overlap_matrices[[1]][upper.tri(overlap_matrices[[1]])]))
overlap_matrices[[2]] <- as.numeric(as.vector(overlap_matrices[[2]][upper.tri(overlap_matrices[[2]])]))
overlap_matrices[[3]] <- as.numeric(as.vector(overlap_matrices[[3]][upper.tri(overlap_matrices[[3]])]))
semantic_net_matrix <- as.numeric(as.vector(semantic_net_matrix[upper.tri(semantic_net_matrix)]))  

seq_vect <- seq(0, 1, 0.1)
par(mfrow=c(2,5))

for(i in 1:9){
  val1 = seq_vect[i]
  j = i + 1
  val2 = seq_vect[j]
  index_of_elements_in_range <- which(semantic_net_matrix > val1 & semantic_net_matrix <= val2)
  chisq_values <- overlap_matrices[[1]][index_of_elements_in_range]
  pvalue_values <- overlap_matrices[[2]][index_of_elements_in_range]
  significant_odds_ratios <-  chisq_values[which(pvalue_values <= 0.00001)]
  significance_text <- paste0(length(significant_odds_ratios),'/', length(chisq_values),  'with p-value <= 10e-5' )
  a = boxplot(as.numeric(as.vector(significant_odds_ratios)), outline=FALSE, main=paste0('Odds ratios at similarities ', val1, '-', val2), col='red', xlab=significance_text, ylim=c(0, 70))
  
  #  hist(as.numeric(as.vector(ciao)))
}
dev.off()
```


We also report the overlap ratios of the couples of annotation classes in the different similarity ranges. The overlap ratios are computed as the number of the LMRs in the intersection of two given classes, divided by their union

```{r plot_overlap_ratios_70_percent, echo=FALSE, eval=FALSE, fig.height=10}

overlap_matrices <- readRDS('./data/ieo_cluster/chisquares_tissue_70_percent_specific.rds')
semantic_net_matrix <- readRDS('./data/semantic_net_matrix_tissues_collapsed.rds')

overlap_matrices[[3]] <- overlap_matrices[[3]][rownames(semantic_net_matrix), colnames(semantic_net_matrix)]
overlap_matrices[[3]] <- as.numeric(as.vector(overlap_matrices[[3]][upper.tri(overlap_matrices[[3]])]))
semantic_net_matrix <- as.numeric(as.vector(semantic_net_matrix[upper.tri(semantic_net_matrix)]))  

png('./images/overlap_tissue_ss_70_percent_specific.png', height=3000, width=4000, res=300)

boxplot(overlap_matrices[[3]] ~ cut(semantic_net_matrix, seq(0,0.9,0.1)), outline=FALSE, xlab='Similarity intervals', ylab='Overlap Rate', main='Rates of overlap of semantic annotation sets',  las=1, varwidth=T)
dev.off()
```

The previous charts show a general increase in the overlap of annotation sets in terms of shared LMRs for similarity values becoming higher. 



####Identification of Semantic set specific LMRs

We decided to use a network approach often used in ecological networks to asses the specialization levels in pollination networks [@dormannspecialist], to determine the specificity of LMRs with respect to semantic annotation sets. By using the package `bipartite` [@dormann2008introducing] we computed different specialization indexes for LMRs in annotation sets.   

<!--
<b>LMRs mean strength</b> is an alternative way to measure the specialization of an LMR for a given semantic class where the proportion of samples annotated with other semantic classes is taken into consideration. For a given LMR the strength is computed as the sum of the weight (number of samples) of each edge connecting the LMR with the annotation classes. If we divide this number by the degree, we obtain the mean number of samples for each LMR. Both the density plots show that most of the LMRs appear in ~75 semantic annotation classes. There is a concentration of LMRs that appear in almost all semantic classes. These LMRs can both be associated to very generic semantic annotation classes or could be themselves generic.

```{r strength_distribution, echo=TRUE, eval=FALSE}
strength <- graph.strength(g, vids=V(g)$type==TRUE)
strength <- strength / deg
summary(strength)
plot(density(strength), xlab='LMRs strength', main='LMRs strength distribution', color='red', lwd=3)
```

The following is a scatterplot of the LMRs degree againts the mean strenght of the LMRs. The green line in the scatterplot represents a regression line, the red lines represent the smoothed conditional
spread, while the central red line represents the non parametric regression smooth. We can see a positive correlation between the degree and mean strength, meaning that higher degree nodes are also supported by a higher number of samples in which the LMRs appear. 

```{r indices, echo=TRUE, eval=FALSE }
#scatterplot(deg, strength, xlab = "LMRs degree", ylab='LMRs mean strength' )
```

In the 'bipartite' package there are several measures to quantify the specificity of LMRs. -->

Using the <b>LMRs degree</b>, the specificity is associated to the number of semantic annotations sets where the LMR has been annotated without considering the number of samples for each LMR. We can say that semantic set specific LMRs can be identified within the LMRs with low low degree (they are associated to few semantic annotation classes) while generic LMRs have higher degrees. In this case, if an LMR appears in n samples, but the 90% of these samples has been annotated within a single semantic annotation set, while the other 10% is distributed overall the other semantic annotation sets, we are saying that the LMR is very generic, even if the evidence for most of the annotation sets is supported by very few samples. 

The <b>strength</b> of a LMR measures the effect of that LMR in a given semantic state. An intermediate matrix is created in the following way: for a given LMR j and a given semantic annotation set i, the value of M[i,j] is given by the number of samples containing LMR j and annotated with semantic annotation set i, divided by the sum of the number of samples of all the other LMRs annotated in semantic annotation set i. The strength of the LMR j is then computed as the sum of the elements in the jth column of the M matrix. A LMR with high strength could be an LMR that appears in many annotation sets and is supported by many samples since it could carry out important functions for all the cells. This doesn't mean necessarily that the LMR is specific.

The <b>species specificity index</b>, also known as coefficient of variation, measures the variation of interactions, where an interaction represents the annotation of a given LMR in a given semantic annotation set. Values close to 0 are associated to low specificity, while values close to 1 are associated to high specificity. 

The <b>pollination service index (PSI)</b> considers boht the strength of the LMRs and of the semantic annotation sets (how much an annotation set is important for a given LMR). 

<b>Partner diversity</b> measures the Shannon entropy of the LMRs, with lower values indicating higher specilization. To compute the partner diversity the bipartite matrix is firstly normalized by columns (dividing each coulmn [LMR] by the sum of weights in that column )  

<b>Effective partners</b> is the partner diversity raised to the power of e.

<b>Proportional generality</b> is the effective partners divided by the number of potential partners.

<b>Proportional similarity</b> measures specializaiton as dissimilarity between between resource use and availability.
Further details about the specificity measures can be found at [@dormann2008introducing]


```{r SPECIFICITY_COMPUTATION, echo=TRUE, eval=FALSE}
ptm <- proc.time()
library(bipartite)
setwd('/data/BA/egaleota/methylation/datasets/')
bipartite_matrix <- readRDS('./bipartite_matrix_tissue_collapsed.rds')
different_specificity_indexes  <- specieslevel(bipartite_matrix, index=c("degree", "species strength", "species specificity", "PSI", "partner diversity", "effective partners","proportional generality", "proportional similarity"), level="higher")
saveRDS(different_specificity_indexes, file='/data/BA/egaleota/methylation/datasets/different_specificity_indexes_tissue.rds')
print(proc.time() - ptm)
```

```{r SPECIFICITY_COMPUTATION_70_percent, echo=TRUE, eval=FALSE}
ptm <- proc.time()
library(bipartite)
setwd('/data/BA/egaleota/methylation/datasets/')
bipartite_matrix <- readRDS('./bipartite_matrix_tissue_collapsed_70_percent.rds')
different_specificity_indexes  <- specieslevel(bipartite_matrix, index=c("degree", "species strength", "species specificity", "PSI", "partner diversity", "effective partners","proportional generality", "proportional similarity"), level="higher")
saveRDS(different_specificity_indexes, file='/data/BA/egaleota/methylation/datasets/different_specificity_indexes_tissue_70_percent.rds')
print(proc.time() - ptm)
```


```{r SPECIFICITY_load, echo=FALSE, eval=FALSE}
different_specificity_indexes <- readRDS('./data/ieo_cluster/different_specificity_indexes_tissue.rds')
different_specificity_indexes_70_percent <- readRDS('./data/ieo_cluster/different_specificity_indexes_tissue_70_percent.rds')
```

In the following chart we show the distribution of specificity scores for the set of LMRs. 

```{r SPECIFICITY_Distribution, echo=FALSE, eval=FALSE, fig.height=12}
png('/Users/egaleota/Desktop/Papers/2017_Onassis/Figures/6_specificities_density.png', res=300, height=2000, width=1000)
par(mfrow=c(4,2))
ciao <- (mapply(function(x,y) plot(density(x), main=y, xlab='Specialization Value'), x=different_specificity_indexes, y=colnames(different_specificity_indexes)))
dev.off()
```


```{r SPECIFICITY_Distribution_70_percent, echo=FALSE, eval=FALSE, fig.height=12}
png('/Users/egaleota/Desktop/Methylation/images/specificity_70_percent.png', res=300, height=2000, width=1000)
par(mfrow=c(4,2))
ciao <- (mapply(function(x,y) plot(density(x), main=y, xlab='Specialization Value'), x=different_specificity_indexes_70_percent, y=colnames(different_specificity_indexes_70_percent)))
dev.off()
```
To compute a threshold for the specificity we generated 1000 null models using the package bipartite. Briefly, each null model assigns to each LMR the same number of semantic annotation sets and the number of samples in which the LMR appears, but scatters the interactions across all semantic annotation sets. These procedure creates LMRs that are extremely generalist. 

```{r nullmodels , echo=TRUE, eval=FALSE}
library(bipartite)
args=commandArgs(trailingOnly=TRUE)
occurrence <- args[1]
bipartite_matrix <- readRDS('./bipartite_matrix_tissue_collapsed.rds')
system.time(chr_null_models <- nullmodel(bipartite_matrix, N=10, method=1))
saveRDS(chr_null_models,  file=paste0('./datasets/nulls_tissue/entire_matrix_model_', as.character(occurrence), '.rds'))
```

```{r nullmodels_70_percent , echo=TRUE, eval=FALSE}
library(bipartite)
args=commandArgs(trailingOnly=TRUE)
occurrence <- args[1]
bipartite_matrix <- readRDS('./bipartite_matrix_tissue_collapsed_70_percent.rds')
system.time(chr_null_models <- nullmodel(bipartite_matrix, N=10, method=1))
saveRDS(chr_null_models,  file=paste0('./datasets/nulls_tissue_70_percent/entire_matrix_model_', as.character(occurrence), '.rds'))
```


We computed the specificity indexes provided in the package bipartite for our null models.

```{r specificity_of_nulls, echo=TRUE, eval=FALSE}

library(bipartite)
setwd('/data/BA/egaleota/methylation/nulls_tissue')
args=commandArgs(trailingOnly=TRUE)
occurrence <- args[1]
ten_models_list <- readRDS(paste0(getwd(), '/entire_matrix_model_', as.character(occurrence), '.rds'))

species_level_indexes <- lapply(ten_models_list, function(single_model){
  all_specificities <- specieslevel(single_model, index=c("degree", "species strength", "species specificity", "PSI", "partner diversity", "effective partners","proportional generality", "proportional similarity"), level="higher")
})
saveRDS(species_level_indexes, file=paste0('/data/BA/egaleota/methylation/datasets/specificities_tissues/specificity_', occurrence, '.rds'))
```


```{r specificity_of_nulls_70, echo=TRUE, eval=FALSE}

library(bipartite)
setwd('/data/BA/egaleota/methylation/nulls_tissue_70_percent')
args=commandArgs(trailingOnly=TRUE)
occurrence <- args[1]
ten_models_list <- readRDS(paste0(getwd(), '/entire_matrix_model_', as.character(occurrence), '.rds'))

species_level_indexes <- lapply(ten_models_list, function(single_model){
  all_specificities <- specieslevel(single_model, index=c("degree", "species strength", "species specificity", "PSI", "partner diversity", "effective partners","proportional generality", "proportional similarity"), level="higher")
})
saveRDS(species_level_indexes, file=paste0('/data/BA/egaleota/methylation/datasets/specificities_tissues_70_percent/specificity_', occurrence, '.rds'))
```


For each of the specificity measures in the package bipartite we obtained the mean over the 1000 null models

```{r distribution_of_specificity_in_null_models, echo=TRUE, eval=FALSE}
setwd('/data/BA/egaleota/methylation/datasets/')
all_specificitiy_files <- list.files('./specificities_tissues_70_percent/', pattern="^specificity.*\\.rds$", full.names=TRUE)
all_specificities <- sapply(all_specificitiy_files, function(specificity_groups){
  group_specificities <- readRDS(specificity_groups)
})
saveRDS(all_specificities, file='./three_dimensional_matrix_of_nulls_specificities_70_percent.rds')
library(plyr)
all_specificity_matrix <- aaply(laply(all_specificities, as.matrix), c(2, 3), mean)
saveRDS(all_specificity_matrix, './null_models_specificity_mean_70_percent.rds')
```


```{r loading_the_null_models_specificities, echo=FALSE, eval=FALSE}
null_models_mean <- readRDS('./data/ieo_cluster/null_models_specificity_mean.rds') 
null_models_mean_70_percent <- readRDS('./data/ieo_cluster/null_models_specificity_mean_70_percent.rds')
```


Finally we plot the distributions of the real network against the one of the null models. The red lines represent real models' specificity distribution.

```{r plot_of_the_specificity_distribution, echo=TRUE, eval=FALSE, fig.height=12}

shannon_spec <- different_specificity_indexes$partner.diversity
limit_x <- max(shannon_spec, null_models_mean[,5])
plot(density(null_models_mean[,5]), col='black', main='Distribution of partner.diversity', xlim=c(0, limit_x))
lines(density(shannon_spec), col='red')

```

```{r plot_of_the_specificity_distribution_70_PERCENT, echo=TRUE, eval=FALSE, fig.height=12}

shannon_spec <- different_specificity_indexes_70_percent$partner.diversity
limit_x <- max(shannon_spec, null_models_mean_70_percent[,5])
plot(density(null_models_mean[,5]), col='black', main='Distribution of partner.diversity', xlim=c(0, limit_x))
lines(density(shannon_spec), col='red')

```



We decided also to apply the threshold solely based on the partner.diversity measure 

```{r thresholding_only_with_partner_diversity, echo=TRUE, eval=FALSE}

thresh_shannon <- quantile(null_models_mean[,5], 0.01)
finals_shannon <- rownames(different_specificity_indexes)[which(different_specificity_indexes[,5]<thresh_shannon)]
saveRDS(finals_shannon, './data/most_specific_lmrs_shannon.rds')

thresh_shannon_70_percent <- quantile(null_models_mean_70_percent[,5], 0.01)
finals_shannon70_percent <- rownames(different_specificity_indexes_70_percent)[which(different_specificity_indexes_70_percent[,5]<thresh_shannon_70_percent)]
saveRDS(finals_shannon, './data/most_specific_lmrs_shannon.rds')
saveRDS(finals_shannon70_percent, './data/most_specific_lmrs_shannon_70_percent.rds')

```

In this case the boxplot of the number of semantic classes for each LMR and vice versa are the following

```{r distribution_of_annotations2, echo=FALSE, eval=FALSE}
unique_edge_list <- readRDS('./data/unique_edge_list_tissue_collapsed.rds')
similarity_specific <- readRDS('./data/most_specific_lmrs_shannon.rds')
unique_edge_list <- unique_edge_list[which(as.character(as.vector(unique_edge_list$lmrs_in_sample)) %in% similarity_specific),]
setDT(unique_edge_list)
par(mfrow=c(1,2))
boxplot(unique_edge_list[, .N, by=list(lmrs_in_sample)]$N, main='Number of semantic classes for each LMR', col='orange', cex.main=0.8)
summary(unique_edge_list[, .N, by=list(lmrs_in_sample)]$N)
boxplot(unique_edge_list[, .N, by=list(annotation)]$N, main='Number of specific LMRs for each semantic class', col='blue', cex.main=0.8)
summary(unique_edge_list[, .N, by=list(annotation)]$N)
```

We computed again overlap ratios of the couples of annotation classes in the different similarity ranges after excluding the general LMRSs. The overlap ratios are computed as the number of the LMRs in the intersection of two given classes, divided by their union

```{r plot_overlap_ratios, echo=FALSE, eval=FALSE, fig.height=10}
#png('./images/figure_6.png', height=3000, width=4000, res=300)

overlap_matrices <- readRDS('./data/ieo_cluster/shannon_specific_overlaps.rds')
semantic_net_matrix <- readRDS('./data/semantic_net_matrix_tissues_collapsed.rds')
overlap_matrices <- overlap_matrices[colnames(semantic_net_matrix),colnames(semantic_net_matrix)]
overlap_matrices <- as.numeric(as.vector(overlap_matrices[upper.tri(overlap_matrices)]))
semantic_net_matrix <- as.numeric(as.vector(semantic_net_matrix[upper.tri(semantic_net_matrix)]))  
boxplot(overlap_matrices ~ cut(semantic_net_matrix, seq(0,0.9,0.1)), outline=FALSE, xlab='Similarity intervals', ylab='Overlap Rate', main='Rates of overlap of semantic annotation sets',  las=1, varwidth=T)
#dev.off()
```


### Functional annotation of Specific LMRs ###

In the work of [@yang2016comparative] authors created an atlas of methylation marks from experimental data. Authors show that most methylation (hyper o hypo regions) segments (~75%) are shared among cell lines, meaning that they are not tissue specific. Most of the low specificity regions were found to be  hypermethylated, while only a very small percentage (~5%) of them  was hypomethylated. The low specificity hypomethylated regions were usually located at CpG islands and gene promoter regions. Furthermore they are usually associated to H3K4me3, while some of them also overlap with ubiquitous enhancers. 
Instead 13% of the methylated regions showed high specificity. Among the highly specific methylated regions some were hypermethylated and other hypomethylated. The methylation profiles of these regions were clustered using k-means and Pearson correlation as distance. Similar groups of cells (different neuron types, different spermatozoa cells etc...) share similar methylation profiles. The enrichment analyses of these regions identified genes whose functions are associated to the specific cell type. 
The paper also shows that cell type specific hypomethylation marks are enriched through H3K27ac and transcription factor binding sites in a cell type specific-manner. 


Based on these results we decided to functionally annotate our distal LMRs. For each semantic set we selected the genes overlapping the LMRs specific for that semantic set in a window of size 500kb (this huge range is due to the fact that we are considering distal LMRs) and performed a functional enrichment analysis obtaining the top GO terms. The following code creates a set of GO annotations for the SHANNON specific LMRs in the semantic sets. 

```{r functional_annotation_of_lmrs, echo=TRUE, eval=FALSE}
#This is the file 08 goAnnotations where we are annotating the granges associated to the LMRs identified as most specific using the shannon diversity
# We save both the topGOres and the Granges annotation to use other tools rather than topgores

library(GenomicRanges)
library(org.Hs.eg.db)
library(parallel)
setwd('/data/BA/egaleota/methylation/')
library(TxDb.Hsapiens.UCSC.hg19.knownGene) 
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene 
library(compEpiTools)
args=commandArgs(trailingOnly=TRUE)
occurrence <- as.numeric(args[1])
file_name <- paste0('/data/BA/egaleota/methylation/datasets/goAnnotationsCollapsed/goAnnotations_', formatC(as.numeric(occurrence), width=3, flag="0"), '.rds')
if(!file.exists(file_name)){
  
  unique_edge_list <- readRDS('./datasets/unique_edge_list_tissue_collapsed.rds')
  
  shannon_specific <- readRDS('./datasets/most_specific_lmrs_shannon.rds')
  
  unique_edge_list <- unique_edge_list[which(as.character(as.vector(unique_edge_list$lmrs_in_sample)) %in% shannon_specific),]
  
  unique_annots <- unique(unique_edge_list$annotation)
  list_of_granges_strings <- mclapply( unique_annots, function(annot){
    lmrs_list <- unique(unique_edge_list$lmrs_in_sample[which(unique_edge_list$annotation==annot)])
  }, mc.cores=12)
  
  names(list_of_granges_strings) <- unique_annots 
  granges <- GRanges(list_of_granges_strings[[occurrence]])
  
  
  centers <- resize(granges, width=1, fix='center')
  
  bed_annotation <- GRannotate(Object=centers, txdb=txdb, EG2GS=org.Hs.eg.db)
  ranges(bed_annotation) <- ranges(granges)
  
  file_name2 <- paste0('/data/BA/egaleota/methylation/datasets/goAnnotationsCollapsed/GRannotate/Annotations_', formatC(as.numeric(occurrence), width=3, flag="0"), '.rds')
  saveRDS(new_granges, file=file_name2)
  gene_symbols <- unique(new_granges$nearest_gene_id) 
  
  
  #As an alternative way one can also use the following code, which when tested gave the same results
  
  #For distal elements such as the ones we are considering we use a window of size 500kb 
  #maxgap is the maximum allowed distance between a query interval and subject interval when doing granges overalp
  #genes_list <- transcriptsByOverlaps(ranges = granges, x = txdb, maxgap=500000, columns=c('tx_id', 'gene_id'))
  #genes <- unique(unlist(genes_list$gene_id))
  new_granges <- bed_annotation[which(bed_annotation$distance_fromTSS<=500000)] 
  
  goannotations <- topGOres(gene_symbols, orgdb=org.Hs.eg.db, Pthr=1e-05) 
  
  
  saveRDS(goannotations, file= file_name)
}
```

We decided to use GREAT tools to annotate directly the genomic regions. The GREAT tool allows to get an enrichment by assigning to each gene in the genome a basal regulatory domain of a minimum distance upstream and downstream of the TSS. The gene regulatory domain is extended in both directions to the nearest gene's basal domain but no more than the maximum extension in one direction. Usually the values used to look upstream the genes' TSSs are 5KB, for downstream 1KB and to consider distal elements the extension is 1Mb. We directly started from the annotated regions obtained by the filtering of LMRs for Shannon specificity score.  

```{r GREAT_annotations_tissue_filtered_70_specific, echo=TRUE, eval=FALSE}
#This file allows to carry out enrichment of the regions 
# in the GRanges previoyusly filtered and annotated with GRannotate of compEpiTools
# and 

setwd('/data/BA/egaleota/methylation/')
library(rGREAT)
args=commandArgs(trailingOnly=TRUE)
occurrence <- as.numeric(args[1])

file_name <- paste0('/data/BA/egaleota/methylation/datasets/GREAT_tissue_70_percent/GREAT_', formatC(as.numeric(occurrence), width=3, flag="0"), '.rds') 
if(! file.exists(file_name))
  
  tissue_2_lmrs_70_percent <- readRDS('./datasets/tissue_2_lmrs_70_percent.rds')
shannon_specific_70_percent <- readRDS('./datasets/most_specific_lmrs_shannon_70_percent.rds')


lmrs_ranges <- tissue_2_lmrs_70_percent[[occurrence]]
lmrs_ranges <- GRanges(lmrs_ranges[which(lmrs_ranges %in% shannon_specific_70_percent)])
#Here we retrieve GRannotate results

#The submitGreatJob function carries out the enrichment of the regions considering basalPlusExt: mode 'Basal plus extension'.
# Gene regulatory domain definition: Each gene is assigned a basal regulatory domain of a minimum distance upstream and downstream of the TSS #(regardless of other nearby genes). The gene regulatory domain is extended in both directions to the nearest gene's basal domain but no more than #the maximum extension in one direction.
#adv_upstream: proximal extension to upstream (unit: kb)  <--- default is 5
#adv_downstream: proximal extension to downstream (unit: kb) <-- default is 1
#adv_span: maximum extension (unit: kb) <- defult is 1000 = 1mb for distal elements
job <- submitGreatJob(lmrs_ranges, species = "hg19")
table_of_annotations <- getEnrichmentTables(job, ontology = availableOntologies(job) )


saveRDS(table_of_annotations, file=file_name)



```


```{r reading_gr_annotate_results_for_shannon_lmrs, echo=TRUE, eval=FALSE}

setwd('/data/BA/egaleota/methylation/')
library(rGREAT)
#Here we retrieve GRannotate results
granges_annotations_list <- lapply(list.files('/data/BA/egaleota/methylation/datasets/goAnnotationsCollapsed/GRannotate/', pattern='*rds*', full.names=TRUE), function(file) readRDS(file))
names(granges_annotations_list) <- list.files('/data/BA/egaleota/methylation/datasets/goAnnotationsCollapsed/GRannotate/', pattern='*rds*')

#We have to determine which pathways are involved in the gene sets obtained from the annotation of Granges associated to LMRs
unique_edge_list <- readRDS('./datasets/unique_edge_list_tissue_collapsed.rds')
unique_edge_list <- unique_edge_list[which(as.character(as.vector(unique_edge_list$lmrs_in_sample)) %in% shannon_specific),]

unique_annots <- unique(unique_edge_list$annotation)
names(granges_annotations_list) <- unique_annots

#The submitGreatJob function carries out the enrichment of the regions considering basalPlusExt: mode 'Basal plus extension'.
# Gene regulatory domain definition: Each gene is assigned a basal regulatory domain of a minimum distance upstream and downstream of the TSS #(regardless of other nearby genes). The gene regulatory domain is extended in both directions to the nearest gene's basal domain but no more than #the maximum extension in one direction.
#adv_upstream: proximal extension to upstream (unit: kb)  <--- default is 5
#adv_downstream: proximal extension to downstream (unit: kb) <-- default is 1
#adv_span: maximum extension (unit: kb) <- defult is 1000 = 1mb for distal elements

granges_job_list <- lapply(granges_annotations_list, function(lmrs_ranges){
  job <- submitGreatJob(lmrs_ranges, species = "hg19")
  table_of_annotations = getEnrichmentTables(job, ontology = availableOntologies(job) )
})

saveRDS(granges_job_list, file='/data/BA/egaleota/methylation/datasets/GREAT_annotations.rds')

```

```{r collapsing_great_tissues_70_percent, echo=FALSE, eval=FALSE}
setwd('/data/BA/egaleota/methylation/')
great_annotations_list <- lapply(list.files('/data/BA/egaleota/methylation/datasets/GREAT_tissue_70_percent/', pattern='*rds*', full.names=TRUE), function(file) readRDS(file))
names(great_annotations_list) <- list.files('/data/BA/egaleota/methylation/datasets/GREAT_tissue_70_percent/', pattern='*rds*')

tissue_2_lmrs_70_percent <- readRDS('./datasets/tissue_2_lmrs_70_percent.rds')

names(great_annotations_list) <- names(tissue_2_lmrs_70_percent) 
saveRDS(great_annotations_list, file='./datasets/GREAT_annotations_tissue_70_percent.rds')
```


Annotations were filtered by the p-value of the binomial distribution test, using as threshold 1e-05. In the following plots we show the annotation of the Shannon specific LMRs in the different semantic sets. The tool allowed the use of different ontologies. Here we report all the results. 


```{r reading_and_filtering_great_annotations, echo=FALSE, eval=FALSE , fig.height= c(10, 40, 10, 40, 10, 30, 5, 20, 15, 5, 5, 22, 40, 40, 20, 10, 19, 15, 5, 20 ),  message=FALSE, warning=FALSE}
great_annotations <- readRDS('./data/ieo_cluster/GREAT_annotations.rds')

pval_filtered_great_annotations <- lapply(great_annotations, function(annotation_list){
  lapply(annotation_list, function(annotation_df){
    annotation_df <- annotation_df[which(annotation_df$Binom_Raw_PValue <= 1e-8),]
  })
})

list_of_matrices <- list()
for(i in 1:21){
  annot_db_name <- names(pval_filtered_great_annotations[[1]])[i]
  list_of_matrices[[i]] <- lapply(pval_filtered_great_annotations, function(semantic_class_annots){
    db_values <- semantic_class_annots[[i]]
    db_values <-db_values[, c(2, 8)]
  })
  names(list_of_matrices)[i] <- annot_db_name
}

unici_valori <- lapply(list_of_matrices, function(measure_name){
  unique(as.character(as.vector(do.call('rbind', lapply(measure_name, '[[', 1)))))
})



mie_matrici <- sapply(names(list_of_matrices), function(matrix_name){
  matrix_of_annotations <- list_of_matrices[[matrix_name]]
  unique_values <- unici_valori[[matrix_name]]
  heatmapDataMatrix <- matrix(0, ncol=length(matrix_of_annotations), nrow=length(unique_values))
  colnames(heatmapDataMatrix) <- names(matrix_of_annotations)
  rownames(heatmapDataMatrix) <- unique_values
  sapply(names(matrix_of_annotations), function(semantic_value_column){
    col_name <- semantic_value_column
    rows <- as.character(as.vector(matrix_of_annotations[[semantic_value_column]]$name))
    heatmapDataMatrix[match(rows, rownames(heatmapDataMatrix)), col_name] <<- -log10(as.numeric(as.vector(matrix_of_annotations[[semantic_value_column]]$Binom_Raw_PValue)))
  })
  heatmapDataMatrix
})


m_a_c <- readRDS('./data/m_a_c_collapsed_semantic_sets.rds')
m_a_c <- unique(m_a_c[,c('nuovi_ids', 'Term_Name')])
m_a_c <- m_a_c[which(!m_a_c$nuovi_ids==""),]
rownames(m_a_c) <- m_a_c$nuovi_ids
col12 <- colorRampPalette(c('white','beige'))
col23 <- colorRampPalette(c('beige','red'))
cols <- c(col12(10), col23(30))



#Alternatively we can plot each matrix separately in a different file
heatmapDataMatrix <- mie_matrici[[1]]
colnames(heatmapDataMatrix) <- m_a_c$Term_Name[match(colnames(heatmapDataMatrix), m_a_c$nuovi_ids)] 
colnames(heatmapDataMatrix) <- gsub("endoderm,", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("mesoderm,", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("ectoderm", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("^,+","", colnames(heatmapDataMatrix))

semantic_net_matrix <- readRDS('./data/semantic_net_matrix_tissues_collapsed.rds')
semantic_distance <- 1 - as.matrix(semantic_net_matrix)
colnames(semantic_distance) <- rownames(semantic_distance) <- m_a_c$Term_Name[match(colnames(semantic_distance), m_a_c$nuovi_ids)]
colnames(semantic_distance) <- gsub("endoderm,", ",", colnames(semantic_distance))
colnames(semantic_distance) <- gsub("mesoderm,", ",", colnames(semantic_distance))
colnames(semantic_distance) <- gsub("ectoderm", ",", colnames(semantic_distance))
colnames(semantic_distance) <- gsub("^,+","", colnames(semantic_distance))
rownames(semantic_distance) <- colnames(semantic_distance)
semantic_distance <- as.dist(semantic_distance)

hclus_semantic_sets_distances <- hclust(d=semantic_distance)
heatmapDataMatrix_c_o <- heatmapDataMatrix[, hclus_semantic_sets_distances$labels]
pheatmap(heatmapDataMatrix_c_o,  cluster_cols=hclus_semantic_sets_distances,  filename='./images/01_GREAT_GO_MF_tissues.png', show_colnames=TRUE, show_rownames = TRUE, width = 12, height=15, fontsize_row = 10, main = 'GO Molecular Function Enrichment')



heatmapDataMatrix <- mie_matrici[[2]]
colnames(heatmapDataMatrix) <- m_a_c$Term_Name[match(colnames(heatmapDataMatrix), m_a_c$nuovi_ids)] 
colnames(heatmapDataMatrix) <- gsub("endoderm,", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("mesoderm,", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("ectoderm", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("^,+","", colnames(heatmapDataMatrix))

heatmapDataMatrix_c_o <- heatmapDataMatrix[, hclus_semantic_sets_distances$labels]

pheatmap(heatmapDataMatrix_c_o,  cluster_cols=hclus_semantic_sets_distances,  filename='./images/02_GREAT_GO_BP_tissues.png', show_colnames=TRUE, show_rownames = TRUE, width = 12, height=35, fontsize_row = 5, fontsize_col = 8, main = 'GO Biological Process Enrichment')


heatmapDataMatrix <- mie_matrici[[3]]
colnames(heatmapDataMatrix) <- m_a_c$Term_Name[match(colnames(heatmapDataMatrix), m_a_c$nuovi_ids)] 
colnames(heatmapDataMatrix) <- gsub("endoderm,", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("mesoderm,", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("ectoderm", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("^,+","", colnames(heatmapDataMatrix))
heatmapDataMatrix_c_o <- heatmapDataMatrix[, hclus_semantic_sets_distances$labels]
pheatmap(heatmapDataMatrix_c_o,  cluster_cols=hclus_semantic_sets_distances,  filename='./images/03_GREAT_GO_CC_tissues.png', show_colnames=TRUE, show_rownames = TRUE, width = 12, height=15, fontsize_row = 10, main = 'GO Cellular Component Enrichment')

heatmapDataMatrix <- mie_matrici[[5]]
colnames(heatmapDataMatrix) <- m_a_c$Term_Name[match(colnames(heatmapDataMatrix), m_a_c$nuovi_ids)] 
colnames(heatmapDataMatrix) <- gsub("endoderm,", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("mesoderm,", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("ectoderm", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("^,+","", colnames(heatmapDataMatrix))
heatmapDataMatrix_c_o <- heatmapDataMatrix[, hclus_semantic_sets_distances$labels]
pheatmap(heatmapDataMatrix_c_o,  cluster_cols=hclus_semantic_sets_distances,  filename='./images/04_GREAT_HumanPhenotype_tissues.png', show_colnames=TRUE, show_rownames = TRUE, width = 12, height=15, fontsize_row = 10, main = 'Human Phenotype Enrichment')

heatmapDataMatrix <- mie_matrici[[6]]
colnames(heatmapDataMatrix) <- m_a_c$Term_Name[match(colnames(heatmapDataMatrix), m_a_c$nuovi_ids)] 
colnames(heatmapDataMatrix) <- gsub("endoderm,", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("mesoderm,", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("ectoderm", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("^,+","", colnames(heatmapDataMatrix))
heatmapDataMatrix_c_o <- heatmapDataMatrix[, hclus_semantic_sets_distances$labels]
pheatmap(heatmapDataMatrix_c_o,  cluster_cols=hclus_semantic_sets_distances,  filename='./images/05_GREAT_DiseaseOntology_tissues.png', show_colnames=TRUE, show_rownames = TRUE, width = 12, height=25, fontsize_row = 10, main = 'Disease Ontology Enrichment')

heatmapDataMatrix <- mie_matrici[[11]]
colnames(heatmapDataMatrix) <- m_a_c$Term_Name[match(colnames(heatmapDataMatrix), m_a_c$nuovi_ids)] 
colnames(heatmapDataMatrix) <- gsub("endoderm,", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("mesoderm,", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("ectoderm", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("^,+","", colnames(heatmapDataMatrix))
heatmapDataMatrix_c_o <- heatmapDataMatrix[, hclus_semantic_sets_distances$labels]
pheatmap(heatmapDataMatrix_c_o,  cluster_cols=hclus_semantic_sets_distances,  filename='./images/06_GREAT_PantherPathway_tissues.png', show_colnames=TRUE, show_rownames = TRUE, width = 10, height=10, fontsize_row = 10, main = 'PANTHER Pathway Enrichment')

heatmapDataMatrix <- mie_matrici[[13]]
colnames(heatmapDataMatrix) <- m_a_c$Term_Name[match(colnames(heatmapDataMatrix), m_a_c$nuovi_ids)] 
colnames(heatmapDataMatrix) <- gsub("endoderm,", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("mesoderm,", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("ectoderm", ",", colnames(heatmapDataMatrix))
colnames(heatmapDataMatrix) <- gsub("^,+","", colnames(heatmapDataMatrix))
heatmapDataMatrix_c_o <- heatmapDataMatrix[, hclus_semantic_sets_distances$labels]
pheatmap(heatmapDataMatrix_c_o,  cluster_cols=hclus_semantic_sets_distances,  filename='./images/07_GREAT_MsIGDB_Pathway_tissues.png', show_colnames=TRUE, show_rownames = TRUE, width = 15, height=20, fontsize_row = 10, main = 'MSigDB Pathway Enrichment')

```




```{r processing_great_tissue_70_percent, echo=FALSE, eval=FALSE , fig.height= c(10, 40, 10, 40, 10, 30, 5, 20, 15, 5, 5, 22, 40, 40, 20, 10, 19, 15, 5, 20 ),  message=FALSE, warning=FALSE}
great_annotations <- readRDS('./data/ieo_cluster/GREAT_annotations_tissue_70_percent.rds')

pval_filtered_great_annotations <- lapply(great_annotations, function(annotation_list){
  lapply(annotation_list, function(annotation_df){
    annotation_df <- annotation_df[which(annotation_df$Binom_Raw_PValue <= 1e-10),]
  })
})



list_of_matrices <- list()
for(i in 1:21){
  annot_db_name <- names(pval_filtered_great_annotations[[1]])[i]
  list_of_matrices[[i]] <- lapply(pval_filtered_great_annotations, function(semantic_class_annots){
    db_values <- semantic_class_annots[[i]]
    db_values <-db_values[, c(1, 2, 8)]
  })
  names(list_of_matrices)[i] <- annot_db_name
}

unici_valori <- lapply(list_of_matrices, function(measure_name){
  unique(unlist(lapply(measure_name, '[[', 2)))
})



mie_matrici <- sapply(names(list_of_matrices), function(matrix_name){
  matrix_of_annotations <- list_of_matrices[[matrix_name]]
  unique_values <- unici_valori[[matrix_name]]
  heatmapDataMatrix <- matrix(0, ncol=length(matrix_of_annotations), nrow=length(unique_values))
  colnames(heatmapDataMatrix) <- names(matrix_of_annotations)
  rownames(heatmapDataMatrix) <- unique_values
  sapply(names(matrix_of_annotations), function(semantic_value_column){
    col_name <- semantic_value_column
    rows <- as.character(as.vector(matrix_of_annotations[[semantic_value_column]]$name))
    heatmapDataMatrix[match(rows, rownames(heatmapDataMatrix)), col_name] <<- -log10(as.numeric(as.vector(matrix_of_annotations[[semantic_value_column]]$Binom_Raw_PValue)))
  })
  heatmapDataMatrix
})


m_a_c <- readRDS('./data/m_a_c_collapsed_semantic_sets.rds')
m_a_c <- unique(m_a_c[,c('nuovi_ids', 'Term_Name')])
m_a_c <- m_a_c[which(!m_a_c$nuovi_ids==""),]
rownames(m_a_c) <- m_a_c$nuovi_ids
col12 <- colorRampPalette(c('white','beige'))
col23 <- colorRampPalette(c('beige','red'))
cols <- c(col12(10), col23(30))



semantic_net_matrix <- readRDS('./data/semantic_net_matrix_tissues_collapsed.rds')
semantic_distance <- 1 - as.matrix(semantic_net_matrix)
colnames(semantic_distance) <- rownames(semantic_distance) <- m_a_c$Term_Name[match(colnames(semantic_distance), m_a_c$nuovi_ids)]
colnames(semantic_distance) <- gsub("endoderm,", ",", colnames(semantic_distance))
colnames(semantic_distance) <- gsub("mesoderm,", ",", colnames(semantic_distance))
colnames(semantic_distance) <- gsub("ectoderm", ",", colnames(semantic_distance))
colnames(semantic_distance) <- gsub("^,+","", colnames(semantic_distance))
rownames(semantic_distance) <- colnames(semantic_distance)
semantic_distance <- as.dist(semantic_distance)

hclus_semantic_sets_distances <- hclust(d=semantic_distance)


heatmapDataMatrix <- mie_matrici[[1]]
goterms <- readRDS('./data/ieo_cluster/goterms.rds')
termini_go <- goterms[match(rownames(heatmapDataMatrix), goterms)]
semplici <- simplifyGOterms(termini_go, maxOverlap = 0.5, ontology = 'MF', go2allEGs = org.Hs.egGO2ALLEGS )
heatmapDataMatrix[match(semplici, rownames(heatmapDataMatrix)),]
heatmapDataMatrix_c_o <- heatmapDataMatrix[, hclus_semantic_sets_distances$labels]
colnames(heatmapDataMatrix_c_o) <- gsub("(.{30,}?)\\s", "\\1\n", colnames(heatmapDataMatrix_c_o))
pheatmap(heatmapDataMatrix_c_o,  cluster_cols=hclus_semantic_sets_distances,  filename='./images/01_GREAT_GO_MF_tissues_70_percent.png', show_colnames=TRUE, show_rownames = TRUE, width = 25, height=40,# fontsize_row = 10, fontsize_col=7,
         main = 'GO Molecular Function Enrichment')



heatmapDataMatrix <- mie_matrici[[2]]
goterms <- readRDS('./data/ieo_cluster/goterms.rds')
termini_go <- goterms[match(rownames(heatmapDataMatrix), goterms)]
semplici <- simplifyGOterms(names(termini_go), maxOverlap = 0.5, ontology = 'BP', go2allEGs = org.Hs.egGO2ALLEGS )
semplici <- semplici[which(!is.na(semplici))]
heatmapDataMatrix <- heatmapDataMatrix[match(termini_go[semplici], rownames(heatmapDataMatrix)),]
heatmapDataMatrix_c_o <- heatmapDataMatrix[, hclus_semantic_sets_distances$labels]
pheatmap(heatmapDataMatrix_c_o,  cluster_cols=hclus_semantic_sets_distances,  filename='./images/02_GREAT_GO_BP_tissues_70_percent.png', show_colnames=TRUE, show_rownames = TRUE, width = 12, height=35, fontsize_row = 5, fontsize_col = 8, main = 'GO Biological Process Enrichment')


heatmapDataMatrix <- mie_matrici[[3]]
heatmapDataMatrix_c_o <- heatmapDataMatrix[, hclus_semantic_sets_distances$labels]
pheatmap(heatmapDataMatrix_c_o,  cluster_cols=hclus_semantic_sets_distances,  filename='./images/03_GREAT_GO_CC_tissues_70_percent.png', show_colnames=TRUE, show_rownames = TRUE, width = 12, height=15, fontsize_row = 10, main = 'GO Cellular Component Enrichment')


heatmapDataMatrix <- mie_matrici[[5]]
heatmapDataMatrix_c_o <- heatmapDataMatrix[, hclus_semantic_sets_distances$labels]
pheatmap(heatmapDataMatrix_c_o,  cluster_cols=hclus_semantic_sets_distances,  filename='./images/04_GREAT_HumanPhenotype_tissues_70_percent.png', show_colnames=TRUE, show_rownames = TRUE, width = 12, height=15, fontsize_row = 10, main = 'Human Phenotype Enrichment')

heatmapDataMatrix <- mie_matrici[[6]]
heatmapDataMatrix_c_o <- heatmapDataMatrix[, hclus_semantic_sets_distances$labels]
pheatmap(heatmapDataMatrix_c_o,  cluster_cols=hclus_semantic_sets_distances,  filename='./images/05_GREAT_DiseaseOntology_tissues_70_percent.png', show_colnames=TRUE, show_rownames = TRUE, width = 12, height=25, fontsize_row = 10, main = 'Disease Ontology Enrichment')

heatmapDataMatrix <- mie_matrici[[11]]
heatmapDataMatrix_c_o <- heatmapDataMatrix[, hclus_semantic_sets_distances$labels]
pheatmap(heatmapDataMatrix_c_o,  cluster_cols=hclus_semantic_sets_distances,  filename='./images/06_GREAT_PantherPathway_tissues_70_percent.png', show_colnames=TRUE, show_rownames = TRUE, width = 10, height=10, fontsize_row = 10, main = 'PANTHER Pathway Enrichment')

heatmapDataMatrix <- mie_matrici[[13]]

heatmapDataMatrix_c_o <- heatmapDataMatrix[, hclus_semantic_sets_distances$labels]
pheatmap(heatmapDataMatrix_c_o,  cluster_cols=hclus_semantic_sets_distances,  filename='./images/07_GREAT_MsIGDB_Pathway_tissues_70_percent.png', show_colnames=TRUE, show_rownames = TRUE, width = 15, height=20, fontsize_row = 10, main = 'MSigDB Pathway Enrichment')

```















### Annotating diseases ###

After the identification of tissue specific LMRs we decided to check wether these LMRs are also specific in diseases. To this end we decided to annotate the samples in Marmalaid with concepts from Human Disease Ontology. The annotation process included some specific rules to identify Healthy samples.  


```{r diseases_annotations, echo=TRUE, eval=FALSE}

ef <- new('EntityFinder')


disease_dict <- new('CMdictionary')
dictRef(disease_dict) <- .jnew('java/io/File', file.path('/Users/egaleota/Library/Mobile Documents/com~apple~CloudDocs/IITDrive/OneDriveFondazioneIstitutoItalianoTecnologia/ChromStates/data/cmDict-doid-non-classified.xml'))
opts <- new('CMoptions')
paramValueIndex(opts) <- 125

ptm <- proc.time()
marmalaid_annotations <- readRDS('./data/marmalaid_annotations.rds')
#Mapping Marmalaid annotations to CL cell lines and DO diseases
marmalaid_annotations$Id <- gsub("-", "_", marmalaid_annotations$Id)

#Annotation of the field DISEASE with DO

marmalaid_annotations$DISEASE <- gsub("^\\s+|\\s+$", "", marmalaid_annotations$DISEASE)
diseases <- unique(marmalaid_annotations$DISEASE)[which(!is.na(unique(marmalaid_annotations$DISEASE)))]
diseases <- diseases[which(!tolower(diseases)=="" & !tolower(diseases)=='na' & !tolower(diseases)=='unknown' & !tolower(diseases)=='maybe')]
diseases <- data.frame(cbind(paste0('disease_', seq(1,length(diseases), 1)), diseases))
marmalaid_disease_entities <- annotateDF(ef, diseases, outDir = './data/a1_files/', configOpt=opts, cmDict=disease_dict)
print(proc.time() - ptm)
marmalaid_disease_entities <- marmalaid_disease_entities[which(!tolower(marmalaid_disease_entities$term_name)=='disease'),]
#Annotation of the field DISEASE_SUBTYPE with DO
marmalaid_annotations$DISEASE_SUBTYPE <- gsub("^\\s+|\\s+$", "", marmalaid_annotations$DISEASE_SUBTYPE)

disease_subtypes <- unique(marmalaid_annotations$DISEASE_SUBTYPE)[which(!is.na(unique(marmalaid_annotations$DISEASE_SUBTYPE)))]
disease_subtypes <- disease_subtypes[-c(71:82)]

disease_subtypes <- disease_subtypes[which(!tolower(disease_subtypes)=="" & !tolower(disease_subtypes)=='na' & !tolower(disease_subtypes)=='unknown' & !tolower(disease_subtypes)=='maybe')]

disease_subtypes <- data.frame(cbind(paste0('diseasesub_', seq(1,length(disease_subtypes), 1)), disease_subtypes))
ptm <- proc.time()
marmalaid_disease_sub_entities <- annotateDF(ef, disease_subtypes, outDir = './data/a1_files/', configOpt=opts, cmDict=disease_dict)
marmalaid_disease_sub_entities <- marmalaid_disease_sub_entities[which(!tolower(marmalaid_disease_sub_entities$term_name)=='disease'),]
print(proc.time()-ptm)

disease_eval <- merge(diseases, unique(marmalaid_disease_entities[,c('sample_id', 'term_name', 'term_url')]), by.x='V1', by.y='sample_id', all.x=TRUE)
colnames(disease_eval) <- c('Term_ID', 'Marmalaid_Annotation', 'Onassis', 'term_url')
disease_eval$Term_ID <- as.character(as.vector(disease_eval$Term_ID))
disease_eval$Marmalaid_Annotation <- as.character(as.vector(disease_eval$Marmalaid_Annotation))
disease_eval$Onassis <- as.character(as.vector(disease_eval$Onassis))
disease_eval$term_url <- as.character(as.vector(disease_eval$term_url))

disease_sub_eval <- merge(disease_subtypes, unique(marmalaid_disease_sub_entities[,c('sample_id', 'term_name', 'term_url')]), by.x='V1', by.y='sample_id', all.x=TRUE)
colnames(disease_sub_eval) <- c('Term_ID', 'Marmalaid_Annotation', 'Onassis', 'term_url')
disease_sub_eval$Term_ID <- as.character(as.vector(disease_sub_eval$Term_ID))
disease_sub_eval$Marmalaid_Annotation <- as.character(as.vector(disease_sub_eval$Marmalaid_Annotation))
disease_sub_eval$Onassis <- as.character(as.vector(disease_sub_eval$Onassis))
disease_sub_eval$term_url <- as.character(as.vector(disease_sub_eval$term_url))


#Merging disease entities
marmalaid_annotations <- merge(marmalaid_annotations, unique(disease_eval[,c('Marmalaid_Annotation', 'term_url', 'Onassis')]), by.x='DISEASE', by.y='Marmalaid_Annotation', all.x=TRUE)
colnames(marmalaid_annotations)[(ncol(marmalaid_annotations)-1):ncol(marmalaid_annotations)] <- c('disease_id', 'disease_name')

#Merging disease_subtype entities
marmalaid_annotations <- merge(marmalaid_annotations, unique(disease_sub_eval[,c('Marmalaid_Annotation', 'term_url', 'Onassis')]), by.x='DISEASE_SUBTYPE', by.y='Marmalaid_Annotation', all.x=TRUE)
colnames(marmalaid_annotations)[(ncol(marmalaid_annotations)-1):ncol(marmalaid_annotations)] <- c('disease_sub_id', 'disease_sub_name')


#Setting healthy samples
marmalaid_annotations$disease_id[which(marmalaid_annotations$DISEASE=='Healthy')] <- 'Healthy'
marmalaid_annotations$disease_name[which(marmalaid_annotations$DISEASE=='Healthy')] <- 'Healthy'
marmalaid_annotations$disease_sub_id[which(marmalaid_annotations$DISEASE=='Healthy')] <- 'Healthy'
marmalaid_annotations$disease_sub_name[which(marmalaid_annotations$DISEASE=='Healthy')] <- 'Healthy'
marmalaid_annotations$disease_id[which(marmalaid_annotations$DISEASE_SUBTYPE=='Healthy')] <- 'Healthy'
marmalaid_annotations$disease_name[which(marmalaid_annotations$DISEASE_SUBTYPE=='Healthy')] <- 'Healthy'
marmalaid_annotations$disease_sub_id[which(marmalaid_annotations$DISEASE_SUBTYPE=='Healthy')] <- 'Healthy'
marmalaid_annotations$disease_sub_name[which(marmalaid_annotations$DISEASE_SUBTYPE=='Healthy')] <- 'Healthy'


saveRDS(marmalaid_annotations, './data/marmalaid_annotations_uncollapsed_disease.rds')
#Creating a data table with marmalaid_annotations

marmalaid_annotations$disease_id <- gsub("^\\,\\s+", "", marmalaid_annotations$disease_id)
marmalaid_annotations$disease_name <- gsub("^\\,\\s+", "", marmalaid_annotations$disease_name)
marmalaid_annotations$disease_sub_id <- gsub("^\\,\\s+", "", marmalaid_annotations$disease_sub_id)
marmalaid_annotations$disease_sub_name <- gsub("^\\,\\s+", "", marmalaid_annotations$disease_sub_name)

setDT(marmalaid_annotations)

#Collapsing the annotations in single annotation lines by sample Id
marmalaid_annotations_collapsed <- marmalaid_annotations[, lapply(.SD, function(x) toString(unique(x[order(x)]))), by=Id]

saveRDS(marmalaid_annotations_collapsed, file='./data/marmalaid_annotations_collapsed_disease.rds')
```

```{r loading_diseases, echo=FALSE, eval=FALSE}
marmalaid_diseases <- readRDS('./data/marmalaid_annotations_collapsed_disease.rds')
```

```{r creating_m_a_c_disease, echo=FALSE, eval=FALSE}
marmalaid_diseases2 <- as.data.frame.matrix(marmalaid_diseases)

#Creation of a data frame where each row contains the annotation ids and annotation names pasted for the first element in the row, that is a geo sample

m_a_c_D <- t(apply(marmalaid_diseases2, 1, function(annotation) {
  id <- annotation[1]
  annot_ids1 <- unique(trimws(strsplit(annotation[14], ",")[[1]]))
  annot_ids2 <- unique(trimws(strsplit(annotation[16], ",")[[1]]))
  annot_ids <- unique(c(annot_ids1, annot_ids2))
  annot_names1 <- unique(trimws(strsplit(annotation[15], ",")[[1]]))
  annot_names2 <- unique(trimws(strsplit(annotation[17], ",")[[1]]))
  annot_names <- unique(c(annot_names1, annot_names2))
  annot_ids <- paste(annot_ids[which(!annot_ids=='NA' & !annot_ids=="" & !is.na(annot_ids))], collapse=',')
  annot_names <- paste(annot_names[which(annot_names!='NA' & annot_names!="" & !is.na(annot_ids))], collapse=',')
  c(id, annot_ids, annot_names)
}))

m_a_c_D[,1] <- gsub('_', '-', m_a_c_D[,1])
rownames(m_a_c_D) <- m_a_c_D[,1]
m_a_c_D <- as.data.frame(m_a_c_D)
colnames(m_a_c_D) <- c('Sample_ID', 'Annot_IDs', 'Annot_Names' )
saveRDS(m_a_c_D, './data/m_a_c_diseases.rds')
saveRDS(m_a_c_D, file='/Users/egaleota/Desktop/Papers/2017_Onassis/Supplementary Files/Supplementary_file3.rds')
```

```{r number_of_samples_for_each_disease_combination, echo=TRUE, eval=FALSE}
setDT(m_a_c_D)
unique_marmalaid_annotations_names <- m_a_c_D[, .N, by=Annot_Names]

```

In the following plot we show the distribution of the number of samples for each semantic disease class. 

```{r boxplot_number_of_samples_for_diseases, echo=FALSE, eval=FALSE}
#Counting the number of samples associated to each semantic annotation class
boxplot(unique_marmalaid_annotations_names$N, main='Number of samples for each combination of disease annotations', col='violet')
summary(unique_marmalaid_annotations_names$N)
```


```{r sample_to_disease_table_loading, echo=FALSE, eval=FALSE}
m_a_c_disease <- readRDS('./data/m_a_c_diseases.rds')
```

For each of the 55 tissue classes obtained in the previous steps we considered the list of samples, their annotated diseases and the set of LMRs associated to them.

```{r diseases_in_annotation_classes, echo=TRUE, eval=FALSE}
#Loading the annotation of tissues
m_a_c_tissue_collapsed <- readRDS('./data/m_a_c_collapsed_semantic_sets.rds')

colnames(m_a_c_disease) <- c("Sample_ID", "Disease_ID", "Disease_Name")
colnames(m_a_c_tissue_collapsed) <- c("Sample_ID", "Tissue_ID", "Tissue_Name")

#Merging the annotation of tissues and diseases
m_a_c <- merge(m_a_c_tissue_collapsed, m_a_c_disease, by='Sample_ID', all.x=TRUE, all.y=TRUE)
m_a_c$Tissue_Name[which(is.na(m_a_c$Tissue_ID))] <- 'Unknown'

m_a_c$Tissue_ID[which(is.na(m_a_c$Tissue_ID))] <- 'Unknown'
m_a_c$Disease_Name <- as.character(as.vector(m_a_c$Disease_Name))
m_a_c$Disease_Name[which(m_a_c$Disease_ID=="")] <- 'Unknown'
m_a_c$Disease_ID <- as.character(as.vector(m_a_c$Disease_ID))
m_a_c$Disease_ID[which(m_a_c$Disease_ID=="")] <- 'Unknown'

unique_tissues <- unique(m_a_c$Tissue_Name)

#Finding the list of GSMs and diseases associated to each tissue
list_of_diseases <- lapply(unique_tissues, function(annotation_ids){
  if(is.na(annotation_ids))
    disease2sm <- unique(m_a_c[which(is.na(m_a_c$Tissue_Name)), c('Sample_ID', 'Disease_Name')])
  else
    diseases2gsm <- unique(m_a_c[which(m_a_c$Tissue_Name==annotation_ids), c('Sample_ID', 'Disease_Name')])
})
names(list_of_diseases) <- unique_tissues

#Finding the list of LMRs associated to each GSM for each disease class in each tissue 
disease_2_lmrs <- lapply(list_of_diseases, function(disease2gsmtable){
  unique_diseases <- unique(disease2gsmtable$Disease_Name)
  disease_lmrs_list <- list()
  if(length(unique_diseases)>=1){
    list_of_lmrs <- list()
    i <- 0
    for(disease in unique_diseases){
      gsms <- disease2gsmtable$Sample_ID[which(disease2gsmtable$Disease_Name==disease)]
      lmrs <- Infi_LMRs_intergenic[gsms]
      i = i + 1
      list_of_lmrs[[i]] <- lmrs
      
      names(list_of_lmrs)[i] <- disease
    }
    disease_lmrs_list <- c(disease_lmrs_list, list_of_lmrs) 
  }
  disease_lmrs_list
})


saveRDS(disease_2_lmrs, file='./data/disease_2_lmrs.rds')



#Unifying the LMRs for a single disease in a single tissue

tissue_2_disease_2_lmrs_no <- lapply(disease_2_lmrs, function(different_diseases_list){
  samples <- lapply(different_diseases_list, function(single_disease_samples){
    all_lmrs <- unlist(lapply(single_disease_samples, as.character))
  }) 
})

#Qui cerco le LMRs consensus per ciascun semantic set combinazione di tessuti e malattie facendo in modo che 
# le low methylated regions in ciascun semantic set siano presenti in almeno il 70 per cento del numero di sample
#che sono annotati con quel tessuto e quella malattia
f_tissue_2_disease_2_lmrs <- lapply(names(tissue_2_disease_2_lmrs_no), function(tissue_name){
  print('Tessuto')
  print(tissue_name)
  tissue_2_disease_list <- tissue_2_disease_2_lmrs_no[[tissue_name]]
  list_of_lmrs <- lapply(names(tissue_2_disease_list), function(disease_name){
    print('Malattia')
    print(disease_name)
    num_sam <- length(disease_2_lmrs[[tissue_name]][[disease_name]])
    print('Numero samples')
    print(num_sam)
    percentage <- round(num_sam * 70 / 100)
    tabella <- table(tissue_2_disease_2_lmrs_no[[tissue_name]][[disease_name]])
    lmrs <- names(which(tabella >= percentage))
  })
  names(list_of_lmrs) <- names(tissue_2_disease_list)
  return(list_of_lmrs)
})

names(f_tissue_2_disease_2_lmrs) <- names(tissue_2_disease_2_lmrs_no)


#tissue_2_disease_2_unique_lmrs <- lapply(disease_2_lmrs, function(different_diseases_list){
#  samples <- lapply(different_diseases_list, function(single_disease_samples){
#    all_lmrs <- unique(unlist(lapply(single_disease_samples, as.character)))
#  }) 
#})

tissue_2_disease_2_unique_lmrs <- f_tissue_2_disease_2_lmrs

saveRDS(tissue_2_disease_2_unique_lmrs, file='./data/tissue_2_disease_2_unique_lmrs.rds')

```

```{r loading_disease_info, echo=FALSE, eval=FALSE}

tissue_2_disease_2_unique_lmrs <- readRDS('./data/tissue_2_disease_2_unique_lmrs.rds')

```

For each tissue class we considered the distinct disease annotations and if there are at least two different disease conditions we compute the overlap rates of the set of LMRs belonging to each diseas/healthy state. The overlap rate of a semantic disease set d1 with a semantic disease set d2 is given by the number of the common LMRs divided by the number of LMRs in d1.


```{r heatmaps_of_overlap, echo=FALSE, eval=FALSE, fig.height=10, fig.width=10}

overl_matrix_list <- lapply(tissue_2_disease_2_unique_lmrs, function(single_tissue_lmrs){
  overl <- NA
  if(length(single_tissue_lmrs)>=2){
    overl <- matrix(100, nrow=length(single_tissue_lmrs), ncol=length(single_tissue_lmrs))
    number_of_diseases <- length(single_tissue_lmrs) 
    rownames(overl) <- colnames(overl) <- names(single_tissue_lmrs)
    for(i in 1:number_of_diseases) {
      lmrs1 <- single_tissue_lmrs[[i]]
      for(j in (1:number_of_diseases)[-i]) {
        lmrs2 <- single_tissue_lmrs[[j]]
        ovl <- 100  * length(which(lmrs1 %in% lmrs2)) / length(lmrs1)
        overl[i, j] <- ovl
      }
    }
  }
  return(overl)
})


names(overl_matrix_list)[which(is.na(names(overl_matrix_list)))] <- 'Unknown'

#myheatmaps <- sapply(names(overl_matrix_list), function(matrix_overlap_name){
#  matrix_overlap <- overl_matrix_list[[matrix_overlap_name]]
#  myheat <- NA
#  if( class(matrix_overlap)=='matrix'){
#    rownames(matrix_overlap)[which(rownames(matrix_overlap)=='')] <- 'Unknown'
#    colnames(matrix_overlap)[which(colnames(matrix_overlap)=='')] <- 'Unknown'
#  
#    cpWB <- colorRampPalette(c('white','beige'))
#    cpBR <- colorRampPalette(c('beige','red'))
#    cols <- c(cpWB(50), cpBR(51)[-1])
#    par(cex.main=0.5)
#    myheat <- heatmap.2(matrix_overlap, cexRow=1.2, cexCol=1.2, col=cols, Rowv=NULL,
#                  Colv=NULL, cellnote=signif(matrix_overlap, 2), 
#                  dendrogram='none', trace='none', density.info='none', notecol='black', #main=matrix_overlap_name, margins=c(25, 25))
#  }
#  return(myheat)
#})


i <- 1
myheatmaps <- sapply(names(overl_matrix_list), function(matrix_overlap_name){
  matrix_overlap <- overl_matrix_list[[matrix_overlap_name]]
  myheat <- NA
  if( class(matrix_overlap)=='matrix' ){
    if('Unknown' %in% rownames(matrix_overlap))
      matrix_overlap <- matrix_overlap[which(!rownames(matrix_overlap)=='Unknown'), which(!colnames(matrix_overlap)=='Unknown') ]
    width_m = nrow(matrix_overlap)
    height_m = ncol(matrix_overlap) 
    main_t = paste0('LMRs overlap in\n', matrix_overlap_name)
    print(matrix_overlap)
    if(class(matrix_overlap)=='matrix'){
      rownames(matrix_overlap) <- gsub("(.{21,}?)\\s", "\\1\n", rownames(matrix_overlap))
      colnames(matrix_overlap) <- rownames(matrix_overlap)
      maint_t <- gsub("(.{21,}?)\\s", "\\1\n", main_t)
      myheat <- pheatmap(matrix_overlap, cluster_rows=FALSE, cluster_cols=FALSE,
                         display_numbers=TRUE, fontsize = 5, fontsize_number=5,
                         number_format = "%.0f", main=main_t, filename = paste0('./images/overl_', as.character(i), '.png'), width=5, height=5 )
      i <<- i + 1
    }
  }
  return(myheat)
})



```

The list of LMRs was then filtered considering only the Shannon specific LMRs. The following images show the overlap matrices considering only specific LMRs. 

```{r specific_lmrs_overlap_in_diseases, echo=TRUE, eval=FALSE, fig.height=10, fig.width=10}

most_specific_lmrs_shannon <- readRDS('./data/most_specific_lmrs_shannon_70_percent.rds')

tissue_2_disease_2_unique_lmrs_specific <- lapply(tissue_2_disease_2_unique_lmrs, function(tissue_lmrs){
  lapply(tissue_lmrs, function(disease_lmrs){
    specific_lmrs <- disease_lmrs[which(disease_lmrs %in% most_specific_lmrs_shannon)]
  })
})



overl_matrix_list <- lapply(tissue_2_disease_2_unique_lmrs_specific, function(single_tissue_lmrs){
  overl <- NA
  lengths <-   as.numeric(as.vector(unlist(lapply(single_tissue_lmrs, length))))
  single_tissue_lmrs <- single_tissue_lmrs[which(lengths>0)]
  if(length(single_tissue_lmrs)>=2){
    overl <- matrix(100, nrow=length(single_tissue_lmrs), ncol=length(single_tissue_lmrs))
    number_of_diseases <- length(single_tissue_lmrs) 
    rownames(overl) <- colnames(overl) <- names(single_tissue_lmrs)
    for(i in 1:number_of_diseases) {
      lmrs1 <- single_tissue_lmrs[[i]]
      for(j in (1:number_of_diseases)[-i]) {
        lmrs2 <- single_tissue_lmrs[[j]]
        ovl <- 100  * length(which(lmrs1 %in% lmrs2)) / length(lmrs1)
        overl[i, j] <- ovl
      }
    }
  }
  return(overl)
})


names(overl_matrix_list)[which(is.na(names(overl_matrix_list)))] <- 'Unknown'

i <- 1
myheatmaps <- sapply(names(overl_matrix_list), function(matrix_overlap_name){
  matrix_overlap <- overl_matrix_list[[matrix_overlap_name]]
  myheat <- NA
  if( class(matrix_overlap)=='matrix' ){
    if('Unknown' %in% rownames(matrix_overlap))
      matrix_overlap <- matrix_overlap[which(!rownames(matrix_overlap)=='Unknown'), which(!colnames(matrix_overlap)=='Unknown') ]
    width_m = nrow(matrix_overlap)
    height_m = ncol(matrix_overlap) 
    main_t = paste0('Specific LMRs overlap in\n', matrix_overlap_name)
    if(class(matrix_overlap)=='matrix'){
      annot_row <- tissue_2_disease_2_unique_lmrs_specific[[matrix_overlap_name]] 
      annot_row <- annot_row[which(!names(annot_row)=='Unknown')]
      nomi <- names(annot_row)
      print(matrix_overlap_name)
      print(nomi)
      if(length(annot_row)>1){
        annot_row <- as.character(sapply(annot_row, length))
        annot_row <- annot_row[which(!annot_row==0)]
        rownames(matrix_overlap) <- paste0(rownames(matrix_overlap), ' (' ,annot_row, ')')
      }
      myheat <- pheatmap(matrix_overlap, cluster_rows=FALSE, cluster_cols=FALSE, display_numbers=TRUE, fontsize = 5, fontsize_number=5, number_format = "%.0f", main=main_t,  filename = paste0('./images/overlap_specific_tissue_disease/overl_specific_', as.character(i), '.png') )
      i <<- i + 1
    }
  }
  return(myheat)
})


#ONLY FOR THE PAPER
i <- 7
matrix_overlap_name <- "brain,brainstem,cerebellum,cerebral hemisphere,cortex,anatomical system,nervous system,central nervous system"
matrix_overlap <- overl_matrix_list[[matrix_overlap_name]]
lunghezze <- lapply(tissue_2_disease_2_unique_lmrs_specific$`brain,brainstem,cerebellum,cerebral hemisphere,cortex,anatomical system,nervous system,central nervous system`, length)
matrix_overlap <- matrix_overlap[which(!rownames(matrix_overlap)=='Unknown'), which(!colnames(matrix_overlap)=='Unknown') ]
lunghezze <- lunghezze[rownames(matrix_overlap)]
rownames(matrix_overlap) <- paste0(rownames(matrix_overlap), '(', unlist(as.numeric(as.vector(lunghezze))), ')')
matrix_overlap <- matrix_overlap[which(lunghezze>16), which(lunghezze>16)]
main_t <- "LMRs overlap in\nbrain,brainstem,cerebellum,cerebral hemisphere,\ncortex,anatomical system,nervous system,central nervous system"
ordine_rows <- hclust(dist(t(matrix_overlap)))
myheat <- pheatmap(matrix_overlap[ordine_rows$order, ordine_rows$order], cluster_rows=FALSE, cluster_cols=TRUE, display_numbers=TRUE, fontsize = 5, fontsize_number=5, number_format = "%.0f", main=main_t,  treeheight_col=15, filename = paste0('./images/overlap_specific_tissue_disease/overl_specific_', as.character(i), '.png'), width=5, height=5)





#ONLY FOR THE PAPER
i <- 15
matrix_overlap_name <- "liver,tube,duct,bile duct"
matrix_overlap <- overl_matrix_list[[matrix_overlap_name]]
lunghezze <- lapply(tissue_2_disease_2_unique_lmrs_specific$`liver,tube,duct,bile duct`, length)
matrix_overlap <- matrix_overlap[which(!rownames(matrix_overlap)=='Unknown'), which(!colnames(matrix_overlap)=='Unknown') ]
lunghezze <- lunghezze[rownames(matrix_overlap)]
rownames(matrix_overlap) <- paste0(rownames(matrix_overlap), '(', unlist(as.numeric(as.vector(lunghezze))), ')')
matrix_overlap <- matrix_overlap[which(lunghezze>16), which(lunghezze>16)]
main_t <- "LMRs overlap in\nliver, tube, duct, bile duct"
ordine_rows <- hclust(dist(t(matrix_overlap)))
myheat <- pheatmap(matrix_overlap[ordine_rows$order, ordine_rows$order], cluster_rows=FALSE, cluster_cols=TRUE, display_numbers=TRUE, fontsize = 5, fontsize_number=5, number_format = "%.0f", main=main_t,  treeheight_col=15, filename = paste0('./images/overlap_specific_tissue_disease/overl_specific_', as.character(i), '.png'), width=5, height=5)



```




```{r specific_lmrs_overlap_in_diseases_tissue_70, echo=TRUE, eval=FALSE, fig.height=10, fig.width=10}

most_specific_lmrs_shannon <- readRDS('./data/most_specific_lmrs_shannon_70_percent.rds')

tissue_2_disease_2_unique_lmrs_specific <- lapply(tissue_2_disease_2_unique_lmrs, function(tissue_lmrs){
  lapply(tissue_lmrs, function(disease_lmrs){
    specific_lmrs <- disease_lmrs[which(disease_lmrs %in% most_specific_lmrs_shannon)]
  })
})

saveRDS(tissue_2_disease_2_unique_lmrs_specific, file='./data/tissue_2_disease_2_lmrs_specific_70.rds')

overl_matrix_list <- lapply(tissue_2_disease_2_unique_lmrs_specific, function(single_tissue_lmrs){
  overl <- NA
  lengths <-   as.numeric(as.vector(unlist(lapply(single_tissue_lmrs, length))))
  single_tissue_lmrs <- single_tissue_lmrs[which(lengths>0)]
  if(length(single_tissue_lmrs)>=2){
    overl <- matrix(100, nrow=length(single_tissue_lmrs), ncol=length(single_tissue_lmrs))
    number_of_diseases <- length(single_tissue_lmrs) 
    rownames(overl) <- colnames(overl) <- names(single_tissue_lmrs)
    for(i in 1:number_of_diseases) {
      lmrs1 <- single_tissue_lmrs[[i]]
      for(j in (1:number_of_diseases)[-i]) {
        lmrs2 <- single_tissue_lmrs[[j]]
        ovl <- 100  * length(which(lmrs1 %in% lmrs2)) / length(lmrs1)
        overl[i, j] <- ovl
      }
    }
  }
  return(overl)
})


names(overl_matrix_list)[which(is.na(names(overl_matrix_list)))] <- 'Unknown'

i <- 1
myheatmaps <- sapply(names(overl_matrix_list), function(matrix_overlap_name){
  matrix_overlap <- overl_matrix_list[[matrix_overlap_name]]
  myheat <- NA
  if( class(matrix_overlap)=='matrix' ){
    if('Unknown' %in% rownames(matrix_overlap))
      matrix_overlap <- matrix_overlap[which(!rownames(matrix_overlap)=='Unknown'), which(!colnames(matrix_overlap)=='Unknown') ]
    width_m = nrow(matrix_overlap)
    height_m = ncol(matrix_overlap) 
    main_t = paste0('Specific LMRs overlap in\n', matrix_overlap_name)
    if(class(matrix_overlap)=='matrix'){
      annot_row <- tissue_2_disease_2_unique_lmrs_specific[[matrix_overlap_name]] 
      annot_row <- annot_row[which(!names(annot_row)=='Unknown')]
      nomi <- names(annot_row)
      print(matrix_overlap_name)
      print(nomi)
      if(length(annot_row)>1){
        annot_row <- as.character(sapply(annot_row, length))
        annot_row <- annot_row[which(!annot_row==0)]
        rownames(matrix_overlap) <- paste0(rownames(matrix_overlap), ' (' ,annot_row, ')')
      }
      myheat <- pheatmap(matrix_overlap, cluster_rows=FALSE, cluster_cols=FALSE, display_numbers=TRUE, fontsize = 5, fontsize_number=5, number_format = "%.0f", main=main_t,  filename = paste0('./images/overlap_specific_tissue_disease/overl_specific_', as.character(i), '.png') , width=5, height=5)
      i <<- i + 1
    }
  }
  return(myheat)
})


#ONLY FOR THE PAPER
i <- 7
matrix_overlap_name <- "brain,brainstem,cerebellum,cerebral hemisphere,cortex,anatomical system,nervous system,central nervous system"
matrix_overlap <- overl_matrix_list[[matrix_overlap_name]]
lunghezze <- lapply(tissue_2_disease_2_unique_lmrs_specific$`brain,brainstem,cerebellum,cerebral hemisphere,cortex,anatomical system,nervous system,central nervous system`, length)
matrix_overlap <- matrix_overlap[which(!rownames(matrix_overlap)=='Unknown'), which(!colnames(matrix_overlap)=='Unknown') ]
lunghezze <- lunghezze[rownames(matrix_overlap)]
rownames(matrix_overlap) <- paste0(rownames(matrix_overlap), '(', unlist(as.numeric(as.vector(lunghezze))), ')')
matrix_overlap <- matrix_overlap[which(lunghezze>16), which(lunghezze>16)]
main_t <- "LMRs overlap in\nbrain,brainstem,cerebellum,cerebral hemisphere,\ncortex,anatomical system,nervous system,central nervous system"
cluster <- hclust(matrix_overlap )
myheat <- pheatmap(matrix_overlap, cluster_rows=FALSE, cluster_cols=FALSE, display_numbers=TRUE, fontsize = 5, fontsize_number=5, number_format = "%.0f", main=main_t,  filename = paste0('./images/overlap_specific_tissue_disease/overl_specific_70_percent', as.character(i), '.png'), width=5, height=5)


```



### GREAT annotation of tissue/disease semantic sets ###

By looking at the overlap matrices for each tissue, we can observe that some LMRs are conserved between different disease classes, while others are lost or acquired in diseas vs healthy. For example let's look at the Myeloyd progenitor, granulocyte....(the first matrix). If we consider all the LMRs we can see that most of the LMRs that are found in the acute promyelotic leukemia are present also in other types of diseases but also in the healthy samples. 
We downloaded the super enhancers of 86 cell lines from dbsuper and verified if our tissue specific low methylated regions overlap with these super enhancers


We decided to annotate the different combinations of tissues/diseases to compare the regions belonging to the different semantic states. 

```{r great_annotation_of_lmrs_based_on_tissues_and_diseases, echo=TRUE, eval=FALSE}
setwd('/data/BA/egaleota/methylation/')
library(rGREAT)
library(GenomicRanges)

args=commandArgs(trailingOnly=TRUE)
occurrence <- as.numeric(args[1])
setwd('/data/BA/egaleota/methylation/')
occurrence <- as.numeric(args[1])


file_name <- paste0('/data/BA/egaleota/methylation/datasets/GREAT_TISSUE_DISEASE/GREAT_',  formatC(as.numeric(occurrence), width=3, flag="0"), '.rds')

if(!file.exists(file_name)){
  
  tissue_2_disease_2_unique_lmrs <- readRDS('./datasets/tissue_2_disease_2_unique_lmrs.rds')
  tissue_2_disease_2_unique_lmrs <- unlist(tissue_2_disease_2_unique_lmrs, recursive = FALSE)
  
  print('OCCORRENZA')
  print(occurrence)
  list_of_lmrs <- tissue_2_disease_2_unique_lmrs[[occurrence]]
  lmrs_ranges <- GRanges(list_of_lmrs)
  print(lmrs_ranges)
  
  
  #The submitGreatJob function carries out the enrichment of the regions considering basalPlusExt: mode 'Basal plus extension'.
  # Gene regulatory domain definition: Each gene is assigned a basal regulatory domain of a minimum distance upstream and downstream of the TSS #(regardless of other nearby genes). The gene regulatory domain is extended in both directions to the nearest gene's basal domain but no more than #the maximum extension in one direction.
  #adv_upstream: proximal extension to upstream (unit: kb)  <--- default is 5
  #adv_downstream: proximal extension to downstream (unit: kb) <-- default is 1
  #adv_span: maximum extension (unit: kb) <- defult is 1000 = 1mb for distal elements
  job <- submitGreatJob(lmrs_ranges, species = "hg19")
  table_of_annotations <- getEnrichmentTables(job, ontology = availableOntologies(job) )
  
  saveRDS(table_of_annotations, file=paste0('/data/BA/egaleota/methylation/datasets/GREAT_TISSUE_DISEASE/GREAT_',  formatC(as.numeric(occurrence), width=3, flag="0"), '.rds'))
}
```

We saved the results of the annotation of all the 21000 LMRs in a single list

```{r unify_great_from_cluster, echo=TRUE, eval=FALSE}
setwd('/data/BA/egaleota/methylation/')
great_annotations_list <- lapply(list.files('/data/BA/egaleota/methylation/datasets/GREAT_TISSUE_DISEASE/', pattern='*rds*', full.names=TRUE), function(file) readRDS(file))
names(great_annotations_list) <- list.files('/data/BA/egaleota/methylation/datasets/GREAT_TISSUE_DISEASE/', pattern='*rds*')
pval_filtered_great_annotations <- lapply(great_annotations_list, function(annotation_list){
  pval_filtered <- lapply(annotation_list, function(annotation_df){
    annotation_df <- annotation_df[which(annotation_df$Binom_Raw_PValue <= 1e-10),]
  })
})
list_of_matrices <- list()
for(i in 1:21){
  annot_db_name <- names(pval_filtered_great_annotations[[1]])[i]
  list_of_matrices[[i]] <- lapply(pval_filtered_great_annotations, function(semantic_class_annots){
    db_values <- semantic_class_annots[[i]]
    db_values <-db_values[, c(2, 8)]
  })
  names(list_of_matrices)[i] <- annot_db_name
}
subsetted_list_of_matrices <- lapply(list_of_matrices, function(annotation_db_ss){
  i = 1 
  lunghezze <- unlist(lapply(tissue_2_disease_2_unique_lmrs, length))
  lista <- list()
  m <- 1
  for(k in lunghezze){
    j = i + k -1
    subset <- annotation_db_ss[i:j]
    i = i + k
    nomi <- names(tissue_2_disease_2_unique_lmrs[[m]])
    names(subset) <- nomi
    lista[[m]] <- subset
    m <- m + 1
  }
  names(lista) <- names(tissue_2_disease_2_unique_lmrs)
  lista
})
list_of_unique_matrices <- lapply(subsetted_list_of_matrices, function(sem_tissue_list){
  tissue_list <- lapply(sem_tissue_list, function(disease_list){
    unici_valori <- unique(unlist(lapply(disease_list, '[[', 1))) 
    heatmapDataMatrix <- matrix(0, nrow=length(unici_valori), ncol=length(disease_list))
    rownames(heatmapDataMatrix) <- unici_valori
    colnames(heatmapDataMatrix) <- names(disease_list)
    filled_heat <- sapply(names(disease_list), function(disease_name){
      rows <- disease_list[[disease_name]][,1]
      col_name <- disease_name
      heatmapDataMatrix[match(rows, rownames(heatmapDataMatrix)), col_name] <<-
        -log10(as.numeric(as.vector(disease_list[[disease_name]][,2])))
    })
    heatmapDataMatrix
  })	
  tissue_list
})
saveRDS(list_of_unique_matrices, file='./datasets/GREAT_annotations_TD.rds')
```

```{r loading_great_annot_of_tissue_and_disease, echo=FALSE, eval=FALSE}
tissue_disease_all_lmrs_greats <- readRDS('./data/ieo_cluster/GREAT_annotations_TD.rds') 
```



Next we plot the heatmaps of great annotation for unfiltered LMRs
```{r creating_heatmaps_of_great_annotations, echo=FALSE, eval=FALSE}
tissue_disease_all_lmrs_greats <- readRDS('./data/ieo_cluster/GREAT_annotations_TD.rds') 
tissue_2_disease_2_unique_lmrs <- readRDS('./data/tissue_2_disease_2_unique_lmrs.rds')
lunghezze <- lapply(tissue_2_disease_2_unique_lmrs, function(diseases) lapply(diseases, length))
i <- 1
matrici <- sapply(names(tissue_disease_all_lmrs_greats), function(database_name){
  db_annot <- tissue_disease_all_lmrs_greats[[database_name]]
  disease_annot <- lapply(names(db_annot), function(tissue_name){
    disease_matrix <- db_annot[[tissue_name]]
    disease_matrix <- disease_matrix[, !colnames(disease_matrix)=='Unknown']
    my_lunghezze <- lunghezze[[tissue_name]]
    print(database_name)
    print(tissue_name)
    
    if(class(disease_matrix)=='matrix'){
      if(ncol(disease_matrix)>1 & nrow(disease_matrix)>0){
        if('Healthy' %in% colnames(disease_matrix)){
          index_of_healthy <- which(colnames(disease_matrix)=='Healthy')
          my_indexes <- rep('FALSE', nrow(disease_matrix))
          for(t in 1:nrow(disease_matrix)){
            fold_change <- abs(disease_matrix[t, index_of_healthy] - disease_matrix[t, -index_of_healthy])
            if(any(fold_change  > 5))
              my_indexes[t] =TRUE
          }
          disease_matrix <- disease_matrix[which(my_indexes==TRUE), ]
          #  print(disease_matrix)
          if(class(disease_matrix)=='matrix'){
            if( ncol(disease_matrix)>1 & nrow(disease_matrix)>0){
              print('COLNAMES')
              print(colnames(disease_matrix))
              lunghezze <- my_lunghezze[colnames(disease_matrix)]
              print('LUGNHEZZE')
              print(lunghezze)
              rownames(disease_matrix) <- gsub("(.{25}?)\\s", "\\1\n", rownames(disease_matrix))
              colnames(disease_matrix) <- gsub("(.{25}?)\\s", "\\1\n", colnames(disease_matrix))
              colnames(disease_matrix) <- paste0(colnames(disease_matrix), ' (', lunghezze , ')' )
              titolo <- gsub("(.{25,}?)\\s", "\\\n", tissue_name)
              titolo <- paste0(titolo, '\n', database_name)
              pheatmap(disease_matrix, main=titolo, file=paste0('./images/GREAT_DISEASES/GR_', i, '.png'), fontsize=4, fontsize_row=3, fontsize_col=4, width=15, height = 15)
              i <<- i + 1
            }}}}}
  })
})

#ONLY FOR THE PAPER
disease_matrix <- tissue_disease_all_lmrs_greats$`GO Biological Process`$`brain,brainstem,cerebellum,cerebral hemisphere,cortex,anatomical system,nervous system,central nervous system`
lunghezze <- lapply(tissue_2_disease_2_unique_lmrs$`brain,brainstem,cerebellum,cerebral hemisphere,cortex,anatomical system,nervous system,central nervous system`, length)
disease_matrix <- disease_matrix[, !colnames(disease_matrix)=='Unknown']
index_of_healthy <- which(colnames(disease_matrix)=='Healthy')
my_indexes <- rep('FALSE', nrow(disease_matrix))
for(t in 1:nrow(disease_matrix)){
  fold_change <- abs(disease_matrix[t, index_of_healthy] - disease_matrix[t, -index_of_healthy])
  if(any(fold_change  > 5))
    my_indexes[t] =TRUE
}
disease_matrix <- disease_matrix[which(my_indexes==TRUE), ]
colnames(disease_matrix) <- paste0(colnames(disease_matrix), ' (', lunghezze[colnames(disease_matrix)], ')')
pheatmap(disease_matrix, main="GO Molecular function Enrichment\n'brain,brainstem,cerebellum,\ncerebral hemisphere,cortex,anatomical system,\nnervous system,central nervous system'", cellwidth=10, cellheight = 10, file='/Users/egaleota/Desktop/Papers/2017_Onassis/Figures/Great_Enrich_brain_No_filter.png', fontsize=4, fontsize_row=5, fontsize_col=5)

#ONLY FOR THE PAPER
lunghezze <- lapply(tissue_2_disease_2_unique_lmrs$`brain,brainstem,cerebellum,cerebral hemisphere,cortex,anatomical system,nervous system,central nervous system`, length)
disease_matrix <- tissue_disease_all_lmrs_greats$`GO Biological Process`$`brain,brainstem,cerebellum,cerebral hemisphere,cortex,anatomical system,nervous system,central nervous system`
disease_matrix <- disease_matrix[, !colnames(disease_matrix)=='Unknown']
index_of_healthy <- which(colnames(disease_matrix)=='Healthy')
my_indexes <- rep('FALSE', nrow(disease_matrix))
for(t in 1:nrow(disease_matrix)){
  fold_change <- abs(disease_matrix[t, index_of_healthy] - disease_matrix[t, -index_of_healthy])
  if(any(fold_change  > 5))
    my_indexes[t] =TRUE
}
disease_matrix <- disease_matrix[which(my_indexes==TRUE), ]
colnames(disease_matrix) <- paste0(colnames(disease_matrix), ' (', lunghezze[colnames(disease_matrix)], ')')
pheatmap(disease_matrix, main="GO Biological Process Enrichment\n'brain,brainstem,cerebellum,\ncerebral hemisphere,cortex,anatomical system,\nnervous system,central nervous system'", cellwidth=20, cellheight = 20, file='/Users/egaleota/Desktop/Papers/2017_Onassis/Figures/10_Great_Enrich_brain_BP.png', fontsize=4, fontsize_row=5, fontsize_col=5)


#ONLY FOR THE PAPER
disease_matrix <- tissue_disease_all_lmrs_greats$`BioCyc Pathway`$`brain,brainstem,cerebellum,cerebral hemisphere,cortex,anatomical system,nervous system,central nervous system`
disease_matrix <- disease_matrix[, !colnames(disease_matrix)=='Unknown']
index_of_healthy <- which(colnames(disease_matrix)=='Healthy')
my_indexes <- rep('FALSE', nrow(disease_matrix))
for(t in 1:nrow(disease_matrix)){
  fold_change <- abs(disease_matrix[t, index_of_healthy] - disease_matrix[t, -index_of_healthy])
  if(any(fold_change  > 2))
    my_indexes[t] =TRUE
}
disease_matrix <- disease_matrix[which(my_indexes==TRUE), ]
pheatmap(disease_matrix, main="GO Biological Process Enrichment\n'brain,brainstem,cerebellum,\ncerebral hemisphere,cortex,anatomical system,\nnervous system,central nervous system'", cellwidth=20, cellheight = 20, file='/Users/egaleota/Desktop/Papers/2017_Onassis/Figures/10_Great_Enrich_brain_pathway.png', fontsize=4, fontsize_row=5, fontsize_col=5)

disease_matrix <- tissue_disease_all_lmrs_greats$`GO Biological Process`$`liver,tube,duct,bile duct`
disease_matrix <- disease_matrix[, !colnames(disease_matrix)=='Unknown']
index_of_healthy <- which(colnames(disease_matrix)=='Healthy')
my_indexes <- rep('FALSE', nrow(disease_matrix))
for(t in 1:nrow(disease_matrix)){
  fold_change <- abs(disease_matrix[t, index_of_healthy] - disease_matrix[t, -index_of_healthy])
  if(any(fold_change  > 20))
    my_indexes[t] =TRUE
}
disease_matrix <- disease_matrix[which(my_indexes==TRUE), ]
rownames(disease_matrix) <- gsub("(.{25?}?)\\s", "\\1\n", rownames(disease_matrix))     colnames(disease_matrix) <- gsub("(.{10}?)\\s", "\\1\n", colnames(disease_matrix))   
colnames(disease_matrix) <- c("Healthy", 
                              "cancer,adenoma,\nhepatocellular adenoma",
                              "cancer,adenoma, carcinoma,\nhepatocellular adenoma",
                              "cancer,adenoma,carcinoma,\nhepatocellular carcinoma",                         "alcohol dependence,\nfatty liver disease, liver disease",
                              "cancer,carcinoma,\nhepatocellular carcinoma",
                              "cancer,cholangiocarcinoma"
)
pheatmap(disease_matrix, main="GO Biological Process Enrichment\n'liver, tube, duct, bile duct'", cellwidth=20, cellheight = 20, file='/Users/egaleota/Desktop/Papers/2017_Onassis/Figures/10_Great_Enrich_TD.png', fontsize=5, fontsize_row=7, fontsize_col=7)


```

The same annotation was carried considering only the list of the most specific LMRs in the different disease/tissue semantic states. 

```{r great_annotation_of_lmrs_based_on_tissues_and_diseases_only_specific_LMRs, echo=TRUE, eval=FALSE}
#This file allows to carry out enrichment of the regions 
# in the GRanges previoyusly filtered and annotated with GRannotate of compEpiTools
# and 

setwd('/data/BA/egaleota/methylation/')
library(rGREAT)
library(GenomicRanges)

args=commandArgs(trailingOnly=TRUE)
occurrence <- as.numeric(args[1])
setwd('/data/BA/egaleota/methylation/')
occurrence <- as.numeric(args[1])

shannon_specific <- readRDS('./datasets/most_specific_lmrs_shannon.rds')
file_name <- paste0('/data/BA/egaleota/methylation/datasets/GREAT_TISSUE_DISEASE_SPEC/GREAT_',  formatC(as.numeric(occurrence), width=3, flag="0"), '.rds')

if(!file.exists(file_name)){
  
  tissue_2_disease_2_unique_lmrs <- readRDS('./datasets/tissue_2_disease_2_unique_lmrs.rds')
  tissue_2_disease_2_unique_lmrs <- unlist(tissue_2_disease_2_unique_lmrs, recursive = FALSE)
  
  print('OCCORRENZA')
  print(occurrence)
  list_of_lmrs <- tissue_2_disease_2_unique_lmrs[[occurrence]]
  list_of_lmrs <- list_of_lmrs[which(list_of_lmrs %in% shannon_specific)]
  lmrs_ranges <- GRanges(list_of_lmrs)
  
  print(lmrs_ranges)
  
  
  #The submitGreatJob function carries out the enrichment of the regions considering basalPlusExt: mode 'Basal plus extension'.
  # Gene regulatory domain definition: Each gene is assigned a basal regulatory domain of a minimum distance upstream and downstream of the TSS #(regardless of other nearby genes). The gene regulatory domain is extended in both directions to the nearest gene's basal domain but no more than #the maximum extension in one direction.
  #adv_upstream: proximal extension to upstream (unit: kb)  <--- default is 5
  #adv_downstream: proximal extension to downstream (unit: kb) <-- default is 1
  #adv_span: maximum extension (unit: kb) <- defult is 1000 = 1mb for distal elements
  job <- submitGreatJob(lmrs_ranges, species = "hg19")
  table_of_annotations <- getEnrichmentTables(job, ontology = availableOntologies(job) )
  
  saveRDS(table_of_annotations, file=paste0('/data/BA/egaleota/methylation/datasets/GREAT_TISSUE_DISEASE_SPEC/GREAT_',  formatC(as.numeric(occurrence), width=3, flag="0"), '.rds'))
}
```

We saved the results of specific lmrs enrichment in a single list

```{r process_shannon_grate, echo=FALSE, eval=FALSE}
setwd('/data/BA/egaleota/methylation/')
great_annotations_list <- lapply(list.files('/data/BA/egaleota/methylation/datasets/GREAT_TISSUE_DISEASE_SPEC/', pattern='*rds*', full.names=TRUE), function(file) readRDS(file))
names(great_annotations_list) <- list.files('/data/BA/egaleota/methylation/datasets/GREAT_TISSUE_DISEASE_SPEC/', pattern='*rds*')



new_annotations_list <- c( great_annotations_list[1:11], NA,
                           great_annotations_list[12:39], NA,
                           
                           great_annotations_list[40:154], NA, NA,
                           great_annotations_list[155], NA, 
                           great_annotations_list[156:184]
)



pval_filtered_great_annotations <- lapply(new_annotations_list, function(annotation_list){
  pval_filtered <- NA
  if(!is.na(annotation_list)){
    pval_filtered <- lapply(annotation_list, function(annotation_df){
      annotation_df <- annotation_df[which(annotation_df$Binom_Raw_PValue <= 1e-5),]
    })
    pval_filtered
  }
  pval_filtered
})




list_of_matrices <- list()
for(i in 1:21){
  annot_db_name <- names(pval_filtered_great_annotations[[1]])[i]
  list_of_matrices[[i]] <- lapply(pval_filtered_great_annotations, function(semantic_class_annots){
    db_values <- NA
    if(class(semantic_class_annots)!='logical'){
      
      if(class(semantic_class_annots[[i]])=='data.frame'){
        if(nrow(semantic_class_annots[[i]])>0){
          db_values <- semantic_class_annots[[i]]
          db_values <-db_values[, c(2, 8)]
        }
      }}
    db_values
    
  })
  names(list_of_matrices)[i] <- annot_db_name
}

tissue_2_disease_2_unique_lmrs <- readRDS('./datasets/tissue_2_disease_2_unique_lmrs.rds')

subsetted_list_of_matrices <- lapply(list_of_matrices, function(annotation_db_ss){
  i = 1 
  lunghezze <- unlist(lapply(tissue_2_disease_2_unique_lmrs, length))
  
  lista <- list()
  m <- 1
  for(k in lunghezze){
    j = i + k -1
    subset <- annotation_db_ss[i:j]
    i = i + k
    nomi <- names(tissue_2_disease_2_unique_lmrs[[m]])
    names(subset) <- nomi
    lista[[m]] <- subset
    
    m <- m + 1
    
  }
  
  names(lista) <- names(tissue_2_disease_2_unique_lmrs)
  
  lista
})






list_of_unique_matrices <- lapply(subsetted_list_of_matrices, function(sem_tissue_list){
  tissue_list <- lapply(sem_tissue_list, function(disease_list){
    disease_list <- disease_list[which(!is.na(disease_list))]
    unici_valori <- unique(unlist(lapply(disease_list, '[[', 1))) 
    heatmapDataMatrix <- matrix(0, nrow=length(unici_valori), ncol=length(disease_list))
    rownames(heatmapDataMatrix) <- unici_valori
    colnames(heatmapDataMatrix) <- names(disease_list)
    filled_heat <- sapply(names(disease_list), function(disease_name){
      rows <- disease_list[[disease_name]][,1]
      col_name <- disease_name
      heatmapDataMatrix[match(rows, rownames(heatmapDataMatrix)), col_name] <<-
        -log10(as.numeric(as.vector(disease_list[[disease_name]][,2])))
    })
    heatmapDataMatrix
  })	
  tissue_list
})
saveRDS(list_of_unique_matrices, file='./datasets/GREAT_annotations_spec.rds')
```

Here we create the heatmaps of GREAT annotation for specific LMRs
```{r creating_heatmaps_of_great_annotations_SPECIFIC, echo=FALSE, eval=FALSE}

tissue_disease_all_lmrs_greats <- readRDS('./data/ieo_cluster/GREAT_annotations_spec_70_percent.rds')
i <- 1
matrici <- sapply(names(tissue_disease_all_lmrs_greats), function(database_name){
  db_annot <- tissue_disease_all_lmrs_greats[[database_name]]
  disease_annot <- lapply(names(db_annot), function(tissue_name){
    disease_matrix <- db_annot[[tissue_name]]
    disease_matrix <- disease_matrix[, !colnames(disease_matrix)=='Unknown']
    
    print(database_name)
    print(tissue_name)
    
    if(class(disease_matrix)=='matrix'){
      if(ncol(disease_matrix)>1 & nrow(disease_matrix)>0){
        if('Healthy' %in% colnames(disease_matrix)){
          index_of_healthy <- which(colnames(disease_matrix)=='Healthy')
          my_indexes <- rep('FALSE', nrow(disease_matrix))
          for(t in 1:nrow(disease_matrix)){
            fold_change <- abs(disease_matrix[t, index_of_healthy] - disease_matrix[t, -index_of_healthy])
            if(any(fold_change  > 5))
              my_indexes[t] =TRUE
          }
          disease_matrix <- disease_matrix[which(my_indexes==TRUE), ]
          print(disease_matrix)
          if(class(disease_matrix)=='matrix'){
            if( ncol(disease_matrix)>1 & nrow(disease_matrix)>0){
              rownames(disease_matrix) <- gsub("(.{25}?)\\s", "\\1\n", rownames(disease_matrix))
              colnames(disease_matrix) <- gsub("(.{25}?)\\s", "\\1\n", colnames(disease_matrix))
              titolo <- gsub("(.{25,}?)\\s", "\\\n", tissue_name)
              titolo <- paste0(titolo, '\n', database_name)
              pheatmap(disease_matrix, main=titolo, file=paste0('./images/GREAT_DISEASES_SPEC/GR_', i, '.png'), fontsize=5, fontsize_row=5, fontsize_col=7)
              i <<- i + 1
            }}}}}
  })
})

#ONLY FOR THE PAPER
lunghezze <- lapply(tissue_2_disease_2_unique_lmrs_specific$`brain,brainstem,cerebellum,cerebral hemisphere,cortex,anatomical system,nervous system,central nervous system`, length)
disease_matrix <- tissue_disease_all_lmrs_greats$`GO Biological Process`$`brain,brainstem,cerebellum,cerebral hemisphere,cortex,anatomical system,nervous system,central nervous system`
disease_matrix <- disease_matrix[, !colnames(disease_matrix)=='Unknown']
index_of_healthy <- which(colnames(disease_matrix)=='Healthy')
my_indexes <- rep('FALSE', nrow(disease_matrix))
for(t in 1:nrow(disease_matrix)){
  fold_change <- abs(disease_matrix[t, index_of_healthy] - disease_matrix[t, -index_of_healthy])
  if(any(fold_change  > 5))
    my_indexes[t] =TRUE
}
disease_matrix <- disease_matrix[which(my_indexes==TRUE), ]
colnames(disease_matrix) <- paste0(colnames(disease_matrix), ' (', lunghezze[colnames(disease_matrix)], ')')
pheatmap(disease_matrix, main="GO Biological Process Enrichment\n'brain,brainstem,cerebellum,\ncerebral hemisphere,cortex,anatomical system,\nnervous system,central nervous system'", cellwidth=20, cellheight = 20, file='/Users/egaleota/Desktop/Papers/2017_Onassis/Figures/08_Great_Enrich_brain_BP.png', fontsize=4, fontsize_row=5, fontsize_col=5)



#ONLY FOR THE PAPER
lunghezze <- lapply(tissue_2_disease_2_unique_lmrs_specific$`liver,tube,duct,bile duct`, length)
disease_matrix <- tissue_disease_all_lmrs_greats$`GO Biological Process`$`liver,tube,duct,bile duct`
disease_matrix <- disease_matrix[, !colnames(disease_matrix)=='Unknown']
index_of_healthy <- which(colnames(disease_matrix)=='Healthy')
my_indexes <- rep('FALSE', nrow(disease_matrix))
for(t in 1:nrow(disease_matrix)){
  fold_change <- abs(disease_matrix[t, index_of_healthy] - disease_matrix[t, -index_of_healthy])
  if(any(fold_change  > 10))
    my_indexes[t] =TRUE
}
disease_matrix <- disease_matrix[which(my_indexes==TRUE), ]
colnames(disease_matrix) <- paste0(colnames(disease_matrix), ' (', lunghezze[colnames(disease_matrix)], ')')
pheatmap(disease_matrix, main="GO Biological Process Enrichment\n'liver, tube, duct, bile duct'", cellwidth=20, cellheight = 10, file='/Users/egaleota/Desktop/Papers/2017_Onassis/Figures/07_Great_Enrich_liver.png', fontsize=4, fontsize_row=5, fontsize_col=5)


```


```{r reading_great_tissue_disease_shannon_70_percent, echo=TRUE, eval=FALSE}
tissue_disease_all_lmrs_greats <- readRDS('./data/ieo_cluster/GREAT_annotations_spec_70_percent.rds')
i <- 1
goterms <- readRDS('./data/ieo_cluster/goterms.rds')

matrici <- sapply(names(tissue_disease_all_lmrs_greats), function(database_name){
  db_annot <- tissue_disease_all_lmrs_greats[[database_name]]
  disease_annot <- lapply(names(db_annot), function(tissue_name){
    disease_matrix <- db_annot[[tissue_name]]
    disease_matrix <- disease_matrix[, !colnames(disease_matrix)=='Unknown']
    
    
    if(class(disease_matrix)=='matrix'){
      if(ncol(disease_matrix)>1 & nrow(disease_matrix)>0){
        if('Healthy' %in% colnames(disease_matrix)){
          index_of_healthy <- which(colnames(disease_matrix)=='Healthy')
          my_indexes <- rep('FALSE', nrow(disease_matrix))
          for(t in 1:nrow(disease_matrix)){
            fold_change <- abs(disease_matrix[t, index_of_healthy] - disease_matrix[t, -index_of_healthy])
            if(any(fold_change  > 5))
              my_indexes[t] =TRUE
          }
          disease_matrix <- disease_matrix[which(my_indexes==TRUE), ]
          if(class(disease_matrix)=='matrix'){
            if( ncol(disease_matrix)>1 & nrow(disease_matrix)>0){
              rownames(disease_matrix) <- gsub("(.{25}?)\\s", "\\1\n", rownames(disease_matrix))
              colnames(disease_matrix) <- gsub("(.{25}?)\\s", "\\1\n", colnames(disease_matrix))
              titolo <- gsub("(.{25,}?)\\s", "\\\n", tissue_name)
              titolo <- paste0(titolo, '\n', database_name)
              w <- 7.5
              h <- 8.65
              if(database_name=='GO Molecular Function'){
                termini_go <- goterms[match(rownames(disease_matrix), goterms)]
                termini_go <- termini_go[which(!is.na(termini_go))]
                semplici <- simplifyGOterms(names(termini_go), maxOverlap = 0.3, ontology = 'MF', go2allEGs = org.Hs.egGO2ALLEGS)
                termini_go <- termini_go[which(names(termini_go) %in% semplici)]
                disease_matrix <- disease_matrix[which(rownames(disease_matrix) %in% termini_go),]
              }
              if(database_name=='GO Biological Process'){
                termini_go <- goterms[match(rownames(disease_matrix), goterms)]
                termini_go <- termini_go[which(!is.na(termini_go))]
                semplici <- simplifyGOterms(names(termini_go), maxOverlap = 0.3, ontology = 'BP', go2allEGs = org.Hs.egGO2ALLEGS)
                termini_go <- termini_go[which(names(termini_go) %in% semplici)]
                disease_matrix <- disease_matrix[which(rownames(disease_matrix) %in% termini_go),]
                
              }
              if(database_name=='GO Cellular Component'){
                termini_go <- goterms[match(rownames(disease_matrix), goterms)]
                termini_go <- termini_go[which(!is.na(termini_go))]
                semplici <- simplifyGOterms(names(termini_go), maxOverlap = 0.3, ontology = 'CC', go2allEGs = org.Hs.egGO2ALLEGS)
                termini_go <- termini_go[which(names(termini_go) %in% semplici)]
                disease_matrix <- disease_matrix[which(rownames(disease_matrix) %in% termini_go),]
                
              }
              print(paste0('i:', i))
              print(dim(disease_matrix))
              
              
              pheatmap(disease_matrix, main=titolo, file=paste0('./images/GREAT_DISEASES_SPEC_70_PERCENT/GR_', i, '.png'),  width=w, height=h, fontsize_row = 3)
              
              i <<- i + 1
            }}}}}
  })
})






#ONLY FOR THE PAPER
lunghezze <- lapply(tissue_2_disease_2_unique_lmrs_specific$`liver,tube,duct,bile duct`, length)
disease_matrix <- tissue_disease_all_lmrs_greats$`GO Biological Process`$`liver,tube,duct,bile duct`
disease_matrix <- disease_matrix[, !colnames(disease_matrix)=='Unknown']
index_of_healthy <- which(colnames(disease_matrix)=='Healthy')
my_indexes <- rep('FALSE', nrow(disease_matrix))
for(t in 1:nrow(disease_matrix)){
  fold_change <- abs(disease_matrix[t, index_of_healthy] - disease_matrix[t, -index_of_healthy])
  if(any(fold_change  > 5))
    my_indexes[t] =TRUE
}
disease_matrix <- disease_matrix[which(my_indexes==TRUE), ]
goterms <- readRDS('./data/ieo_cluster/goterms.rds')
termini_go <- goterms[match(rownames(disease_matrix), goterms)]
termini_go <- termini_go[which(!is.na(termini_go))]
semplici <- simplifyGOterms(names(termini_go), maxOverlap = 0.3, ontology = 'BP', go2allEGs = org.Hs.egGO2ALLEGS )
specifici_liver <- rownames(tissue_disease_all_lmrs_greats$`GO Biological Process`$`liver,tube,duct,bile duct`)[grep('liver|bile|duct|hepat|tube|fat|keton|lipid|amin|gluco|insul|lipo|adipo|neur|hormo', tolower(rownames(tissue_disease_all_lmrs_greats$`GO Biological Process`$`liver,tube,duct,bile duct`)))]
termini_go <- termini_go[which(names(termini_go) %in% semplici )]

disease_matrix <- disease_matrix[which(rownames(disease_matrix) %in% termini_go | rownames(disease_matrix) %in% specifici_liver),]

colnames(disease_matrix) <- paste0(colnames(disease_matrix), ' (', lunghezze[colnames(disease_matrix)], ')')
pheatmap(disease_matrix, main="GO Biological Process Enrichment\n'liver, tube, duct, bile duct'", file='/Users/egaleota/Desktop/Papers/2017_Onassis/Figures/07_Great_Enrich_liver.png', fontsize=4, fontsize_row=3, fontsize_col=4, width=5,height = 7, treeheight_row=15, treeheight_col=15)





lunghezze <- lapply(tissue_2_disease_2_unique_lmrs_specific$`breast,mammary gland,gland,thoracic mammary gland`, length)
disease_matrix <- tissue_disease_all_lmrs_greats$`GO Biological Process`$`breast,mammary gland,gland,thoracic mammary gland`
disease_matrix <- disease_matrix[, !colnames(disease_matrix)=='Unknown']
index_of_healthy <- which(colnames(disease_matrix)=='Healthy')
my_indexes <- rep('FALSE', nrow(disease_matrix))
for(t in 1:nrow(disease_matrix)){
  fold_change <- abs(disease_matrix[t, index_of_healthy] - disease_matrix[t, -index_of_healthy])
  if(any(fold_change  > 5))
    my_indexes[t] =TRUE
}
disease_matrix <- disease_matrix[which(my_indexes==TRUE), ]
#goterms <- readRDS('./data/ieo_cluster/goterms.rds')
#termini_go <- goterms[match(rownames(disease_matrix), goterms)]
#termini_go <- termini_go[which(!is.na(termini_go))]
#semplici <- simplifyGOterms(names(termini_go), maxOverlap = 0.3, ontology = 'BP', go2allEGs = org.Hs.egGO2ALLEGS )
#termini_go <- termini_go[which(names(termini_go) %in% semplici)]
#disease_matrix <- disease_matrix[which(rownames(disease_matrix) %in% termini_go),]

colnames(disease_matrix) <- paste0(colnames(disease_matrix), ' (', lunghezze[colnames(disease_matrix)], ')')
pheatmap(disease_matrix, main="GO Biological Process Enrichment\n'breast'", file='/Users/egaleota/Desktop/Papers/2017_Onassis/Figures/Breast.png', fontsize=4, fontsize_row=4, fontsize_col=4, width=5,height = 10, treeheight_row=15, treeheight_col=15)












#ONLY FOR THE PAPER
lunghezze <- lapply(tissue_2_disease_2_unique_lmrs_specific$`brain,brainstem,cerebellum,cerebral hemisphere,cortex,anatomical system,nervous system,central nervous system`, length)
disease_matrix <- tissue_disease_all_lmrs_greats$`GO Biological Process`$`brain,brainstem,cerebellum,cerebral hemisphere,cortex,anatomical system,nervous system,central nervous system`
disease_matrix <- disease_matrix[, !colnames(disease_matrix)=='Unknown']
index_of_healthy <- which(colnames(disease_matrix)=='Healthy')
my_indexes <- rep('FALSE', nrow(disease_matrix))
for(t in 1:nrow(disease_matrix)){
  fold_change <- abs(disease_matrix[t, index_of_healthy] - disease_matrix[t, -index_of_healthy])
  if(any(fold_change  > 5))
    my_indexes[t] =TRUE
}
disease_matrix <- disease_matrix[which(my_indexes==TRUE), ]
goterms <- readRDS('./data/ieo_cluster/goterms.rds')
termini_go <- goterms[match(rownames(disease_matrix), goterms)]
termini_go <- termini_go[which(!is.na(termini_go))]
semplici <- simplifyGOterms(names(termini_go), maxOverlap = 0.3, ontology = 'BP', go2allEGs = org.Hs.egGO2ALLEGS )
termini_go <- termini_go[which(names(termini_go) %in% semplici)]
disease_matrix <- disease_matrix[which(rownames(disease_matrix) %in% termini_go),]


colnames(disease_matrix) <- paste0(colnames(disease_matrix), ' (', lunghezze[colnames(disease_matrix)], ')')
pheatmap(disease_matrix, main="GO Biological Process Enrichment\n'brain,brainstem,cerebellum,\ncerebral hemisphere,cortex,anatomical system,\nnervous system,central nervous system'", #cellwidth=10, cellheight = 10,
         file='/Users/egaleota/Desktop/Papers/2017_Onassis/Figures/08_Great_Enrich_brain_BP_backup.png', fontsize=4, fontsize_row=4, fontsize_col=5, width=5,height = 16, treeheight_row=15, treeheight_col=15)



pheatmap(tissue_disease_all_lmrs_greats$`GO Biological Process`$`brain,brainstem,cerebellum,cerebral hemisphere,cortex,anatomical system,nervous system,central nervous system`[grep('neur|brain|cereb|nerv|cran', tolower(rownames(tissue_disease_all_lmrs_greats$`GO Biological Process`$`brain,brainstem,cerebellum,cerebral hemisphere,cortex,anatomical system,nervous system,central nervous system`))),])





pheatmap(tissue_disease_all_lmrs_greats$`GO Biological Process`$`liver,tube,duct,bile duct`[grep('liver|bile|duct|hepat|aden|carcin|chola|tube|neph|metabolic|keton|wound|lipid|amin', tolower(rownames(tissue_disease_all_lmrs_greats$`GO Biological Process`$`liver,tube,duct,bile duct`))),])


pheatmap(tissue_disease_all_lmrs_greats$`GO Biological Process`$`breast,mammary gland,gland,thoracic mammary gland`[grep('breast|mamm|gland|thor|horm', tolower(rownames(tissue_disease_all_lmrs_greats$`GO Biological Process`$`breast,mammary gland,gland,thoracic mammary gland`))),])

pheatmap(tissue_disease_all_lmrs_greats$`GO Biological Process`$`blood,B cell,naive B cell,T cell`[grep('blood|leuko|lympho|monon|B cell|vascul|hemato' , tolower(rownames(tissue_disease_all_lmrs_greats$`GO Biological Process`$`blood,B cell,naive B cell,T cell`))),]) 

```

```{r overall_overlap_in_189_tissues, echo=FALSE, eval=FALSE}
tissue_2_disease_2_unique_lmrs <- unlist(readRDS('./data/tissue_2_disease_2_unique_lmrs.rds'), recursive = FALSE)

overl_matrix <- matrix(100, length(tissue_2_disease_2_unique_lmrs), length(tissue_2_disease_2_unique_lmrs))

rownames(overl_matrix) <- colnames(overl_matrix) <- names(tissue_2_disease_2_unique_lmrs)

for(i in 1:length(tissue_2_disease_2_unique_lmrs)){
  print(i)
  lmrs1 <- tissue_2_disease_2_unique_lmrs[[i]]
  for(j in 1:length(tissue_2_disease_2_unique_lmrs)){
    lmrs2 <- tissue_2_disease_2_unique_lmrs[[j]]
    ovl <- 100  * length(which(lmrs1 %in% lmrs2)) / length(lmrs1)
    overl_matrix[i, j] <- ovl
  }
}
ordine_rows <- hclust(dist(t(overl_matrix)))
overl_matrix2 <- overl_matrix[ordine_rows$order, ordine_rows$order]
lunghezze <- unlist(lapply(tissue_2_disease_2_unique_lmrs, length), recursive=FALSE)
lunghezze <- lunghezze[rownames(overl_matrix2)]
rownames(overl_matrix2) <- paste0(rownames(overl_matrix2), ' (', lunghezze, ')')
pheatmap(overl_matrix2, cluster_rows=TRUE, cluster_cols=TRUE, display_numbers=TRUE, fontsize = 4, fontsize_number=2, number_format = "%.0f", main='Overall overlap',  filename = paste0('./images/overlap_specific_tissue_disease/overl_overall.png'), width=15, height=15, treeheight_col=50 )


```
# References
